// @mindfiredigital/pivothead/dist/pivothead-core.mjs
import * as x from "xlsx";
import { jsPDF as pe } from "jspdf";
import { autoTable as we } from "jspdf-autotable";
function re(c, e, t) {
  if (!c || c.length === 0) return 0;
  const r = c.map((s) => Number(s[e]) || 0);
  switch (t) {
    case "sum":
      return r.reduce((s, a) => s + a, 0);
    case "avg":
      return r.reduce((s, a) => s + a, 0) / r.length;
    case "min":
      return Math.min(...r);
    case "max":
      return Math.max(...r);
    case "count":
      return r.length;
    default:
      return 0;
  }
}
function ye(c, e, t) {
  return !e || e.length === 0 ? c : [...c].sort((r, s) => {
    for (const a of e) {
      const { field: i, direction: n, type: o, aggregation: u } = a;
      if (o === "measure") {
        const d = se(r, i), h = se(s, i);
        if (d !== h)
          return n === "asc" ? d - h : h - d;
      } else {
        const d = r[i], h = s[i];
        if (d !== h)
          return n === "asc" ? d < h ? -1 : 1 : d > h ? -1 : 1;
      }
    }
    return 0;
  });
}
function se(c, e, t) {
  return Number(c[e]) || 0;
}
function be(c, e = null, t = null) {
  let r = [...c.data || []];
  e && (r = ye(r, [e]));
  let s = [];
  if (t) {
    const { rowFields: a, columnFields: i, grouper: n } = t, o = [...a, ...i];
    s = ce(r, o, n);
  }
  return { rawData: r, groups: s };
}
function ce(c, e, t) {
  if (!e || e.length === 0)
    return [{ key: "All", items: c, subgroups: [], aggregates: {} }];
  const r = {};
  return c.forEach((s) => {
    const a = t(s, e);
    r[a] || (r[a] = { key: a, items: [], subgroups: [], aggregates: {} }), r[a].items.push(s);
  }), e.length > 1 && Object.values(r).forEach((s) => {
    s.subgroups = ce(
      s.items,
      e.slice(1),
      t
    );
  }), Object.values(r);
}
const Se = "@mindfiredigital/pivothead", ue = 1 * 1024 * 1024, j = 5 * 1024 * 1024, q = 8 * 1024 * 1024, B = 10 * 1024 * 1024, Ce = 256 * 1024, ae = 1 * 1024 * 1024, De = 2 * 1024 * 1024, ve = 5 * 1024 * 1024, ke = 10 * 1024 * 1024, Fe = 100 * 1024 * 1024, Ne = 100, Ee = [
  "region",
  "country",
  "state",
  "city",
  "category",
  "department"
], Ae = [
  "product",
  "item",
  "month",
  "quarter",
  "year",
  "type"
], Me = 40, _e = 10, Le = 50, Re = 10, xe = "#f8f9fa", de = Se, K = {
  error: 0,
  warn: 1,
  info: 2,
  debug: 3
};
function Pe() {
  const c = typeof process < "u" ? process.env.LOG_LEVEL ?? "info" : "info";
  return K[c] ?? K.info;
}
function $e() {
  const c = Pe(), e = (t, r, s, ...a) => {
    (K[r] ?? 99) <= c && t(`[${de}] ${r.toUpperCase()}: ${s}`, ...a);
  };
  return {
    error: (t, ...r) => e(console.error.bind(console), "error", t, ...r),
    warn: (t, ...r) => e(console.warn.bind(console), "warn", t, ...r),
    info: (t, ...r) => e(console.info.bind(console), "info", t, ...r),
    debug: (t, ...r) => e(console.debug.bind(console), "debug", t, ...r)
  };
}
function Te() {
  if (typeof window < "u") return null;
  try {
    const c = require("winston"), { combine: e, label: t, colorize: r, printf: s, timestamp: a } = c.format, i = s(
      ({
        level: n,
        message: o,
        label: u,
        timestamp: d,
        ...h
      }) => {
        const g = Object.keys(h).length ? ` ${JSON.stringify(h)}` : "";
        return `${String(d)} [${String(u)}] ${String(n)}: ${String(o)}${g}`;
      }
    );
    return c.createLogger({
      level: (typeof process < "u" ? process.env.LOG_LEVEL : void 0) ?? "info",
      format: e(
        t({ label: de }),
        a({ format: "YYYY-MM-DD HH:mm:ss" }),
        r({ all: !1 }),
        i
      ),
      transports: [new c.transports.Console()]
    });
  } catch {
    return null;
  }
}
const l = Te() ?? $e();
function P(c) {
  return (c == null ? "" : String(c)).replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#039;");
}
class E {
  /**
   * Converts the pivot table data to HTML
   * @param {PivotTableState<T>} state - The current state of the pivot table
   * @returns {string} HTML string representation of the pivot table
   */
  static convertToHtml(e) {
    const { rows: t, columns: r, selectedMeasures: s, formatting: a, rawData: i } = e;
    if (i.length === 0 || !t.length || !r.length)
      return "<div>No data to display</div>";
    const n = [
      ...new Set(i.map((g) => g[r[0].uniqueName]))
    ], o = [
      ...new Set(i.map((g) => g[t[0].uniqueName]))
    ], u = (g, m) => g === 0 ? "$0.00" : !g && g !== 0 ? "" : m && m.type === "currency" ? new Intl.NumberFormat(m.locale, {
      style: "currency",
      currency: m.currency,
      minimumFractionDigits: m.decimals,
      maximumFractionDigits: m.decimals
    }).format(g) : m && m.type === "number" ? new Intl.NumberFormat(m.locale, {
      minimumFractionDigits: m.decimals,
      maximumFractionDigits: m.decimals
    }).format(g) : String(g), d = (g, m) => "";
    let h = `
  <div class="pivot-export">
    <style>
      .pivot-table {
        border-collapse: collapse;
        width: 100%;
        font-family: Arial, sans-serif;
      }
      .pivot-table th, .pivot-table td {
        border: 1px solid #ddd;
        padding: 8px;
        text-align: right;
      }
      .pivot-table th {
        background-color: #f2f2f2;
        font-weight: bold;
        text-align: center;
        position: relative;
      }
      .pivot-table .sort-icon::after {
        content: "↕";
        position: absolute;
        right: 4px;
        opacity: 0.5;
      }
      .pivot-table th.region-header {
        border-bottom: none;
      }
      .pivot-table th.measure-header {
        border-top: none;
      }
      .pivot-table .row-header {
        text-align: left;
        font-weight: bold;
        background-color: #f9f9f9;
      }
      .pivot-table .corner-header {
        background-color: #f2f2f2;
        border-bottom: 1px solid #ddd;
      }
      .pagination {
        margin-top: 15px;
        font-family: Arial, sans-serif;
      }
      .export-info {
        margin-top: 15px;
        font-size: 0.8em;
        color: #666;
        font-family: Arial, sans-serif;
      }
    </style>

    <table class="pivot-table">
      <thead>
        <tr>
          <th rowspan="2" class="corner-header">${P(
      t[0]?.caption || t[0]?.uniqueName || ""
    )} /<br>Region</th>`;
    return n.forEach((g) => {
      h += `<th colspan="${s.length}" class="region-header">${P(g)}</th>`;
    }), h += `
        </tr>
        <tr>`, n.forEach(() => {
      s.forEach((g) => {
        h += `<th class="measure-header sort-icon">${P(
          g.caption || g.uniqueName
        )}</th>`;
      });
    }), h += `
        </tr>
      </thead>
      <tbody>`, o.forEach((g) => {
      h += `<tr>
      <td class="row-header">${P(g)}</td>`, n.forEach((m) => {
        s.forEach((C) => {
          const S = i.filter(
            (w) => w[t[0].uniqueName] === g && w[r[0].uniqueName] === m
          );
          let b = 0;
          if (S.length > 0)
            switch (C.aggregation) {
              case "sum":
                b = S.reduce(
                  (w, y) => w + (Number(y[C.uniqueName]) || 0),
                  0
                );
                break;
              case "avg":
                C?.formula && typeof C.formula == "function" ? b = S.reduce(
                  (w, y) => w + (C.formula?.(y) || 0),
                  0
                ) / S.length : b = S.reduce(
                  (w, y) => w + (Number(y[C.uniqueName]) || 0),
                  0
                ) / S.length;
                break;
              case "max":
                b = Math.max(
                  ...S.map(
                    (w) => Number(w[C.uniqueName]) || 0
                  )
                );
                break;
              case "min":
                b = Math.min(
                  ...S.map(
                    (w) => Number(w[C.uniqueName]) || 0
                  )
                );
                break;
              case "count":
                b = S.length;
                break;
              default:
                b = 0;
            }
          const f = d(b, C.uniqueName), p = u(
            b,
            a[C.uniqueName]
          );
          h += `<td${f}>${P(p)}</td>`;
        });
      }), h += "</tr>";
    }), h += `
      </tbody>
    </table>

    <div class="pagination">
      Page ${e.paginationConfig?.currentPage || 1} of ${e.paginationConfig?.totalPages || 1}
    </div>

    <div class="export-info">
      <p>Generated: ${(/* @__PURE__ */ new Date()).toLocaleString()}</p>
    </div>
  </div>`, h;
  }
  /**
   * Exports the pivot table data to HTML and downloads the file
   * @param {PivotTableState<T>} state - The current state of the pivot table
   * @param {string} fileName - The name of the downloaded file (without extension)
   */
  static exportToHTML(e, t = "pivot-table") {
    l.info(
      "PivotExportService.exportToHTML called with fileName:",
      t
    ), l.info("State rawData length:", e.rawData?.length || 0);
    const r = E.convertToHtml(e);
    l.info("HTML content length:", r.length);
    const s = `<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
  <title>${P(t)}</title>
  <style>
    body { font-family: Arial, sans-serif; margin: 20px; }
    .pivot-export { max-width: 100%; overflow-x: auto; }
  </style>
</head>
<body>
  ${r}
</body>
</html>`;
    l.info("Full HTML length:", s.length);
    const a = "data:text/html;charset=utf-8," + encodeURIComponent(s);
    l.info("Data URL created, length:", a.length);
    const i = document.createElement("a");
    i.href = a, i.download = `${t}.html`, document.body.appendChild(i), l.info("Clicking download link..."), i.click(), document.body.removeChild(i), l.info("HTML export completed");
  }
  /**
   * Exports the pivot table data to PDF and downloads the file
   * @param {PivotTableState<T>} state - The current state of the pivot table
   * @param {string} fileName - The name of the downloaded file (without extension)
   */
  static exportToPDF(e, t = "pivot-table") {
    l.info(
      "PivotExportService.exportToPDF called with fileName:",
      t
    ), l.info("State rawData length:", e.rawData?.length || 0);
    const r = E.convertToHtml(e);
    l.info("HTML content length for PDF:", r.length);
    const s = document.createElement("div");
    s.style.position = "absolute", s.style.left = "-9999px", s.innerHTML = r, document.body.appendChild(s);
    const a = s.querySelector("table");
    if (!a) {
      l.error("No table found in the generated HTML"), document.body.removeChild(s);
      return;
    }
    l.info("Table element found, proceeding with PDF generation");
    try {
      l.info("Creating jsPDF instance...");
      const i = new pe();
      i.setFontSize(16), i.text(t, i.internal.pageSize.getWidth() / 2, 15, {
        align: "center"
      }), we(i, {
        html: a,
        startY: 25,
        styles: {
          fontSize: 10,
          cellPadding: 3,
          overflow: "linebreak"
        },
        headStyles: {
          fillColor: [66, 139, 202],
          textColor: 255,
          fontStyle: "bold"
        },
        columnStyles: {},
        margin: { top: 25, right: 15, bottom: 25, left: 15 },
        didDrawPage: (n) => {
          i.setFontSize(10), i.text(
            `Page ${n.pageNumber}`,
            i.internal.pageSize.getWidth() - 20,
            i.internal.pageSize.getHeight() - 10
          );
        }
      }), i.save(`${t}.pdf`), document.body.removeChild(s);
    } catch (i) {
      l.error("Error exporting to PDF:", i), document.body.removeChild(s);
    }
  }
  /**
   * Exports the pivot table data to Excel and downloads the file
   * @param {PivotTableState<T>} state - The current state of the pivot table
   * @param {string} fileName - The name of the downloaded file (without extension)
   */
  static exportToExcel(e, t = "pivot-table") {
    l.info(
      "PivotExportService.exportToExcel called with fileName:",
      t
    ), l.info("State rawData length:", e.rawData?.length || 0);
    try {
      E.generateExcel(e, t);
    } catch (r) {
      l.error("Error exporting to Excel:", r);
    }
  }
  /**
   * Generates an Excel file from the pivot table data
   * @param {PivotTableState<T>} state - The current state of the pivot table
   * @param {string} fileName - The name of the downloaded file (without extension)
   */
  static generateExcel(e, t) {
    if (!e.data || e.data.length === 0) {
      l.info("No data to export!");
      return;
    }
    const r = e.rows || [], s = e.columns || [], a = e.measures || [], i = r[0]?.uniqueName, n = s[0]?.uniqueName;
    if (!i || !n) {
      l.info("Missing row or column dimension");
      return;
    }
    const o = [
      ...new Set(e.data.map((f) => f[i]))
    ], u = [
      ...new Set(e.data.map((f) => f[n]))
    ], d = [
      r[0]?.caption || "Dimension",
      ...u.flatMap(
        (f) => a.map(
          (p) => `${f} - ${p.caption || p.uniqueName}`
        )
      )
    ], h = o.map((f) => {
      const p = [f];
      return u.forEach((w) => {
        a.forEach((y) => {
          const D = e.data.filter(
            (v) => v[i] === f && v[n] === w
          );
          let k = 0;
          if (D.length > 0)
            switch (y.aggregation) {
              case "sum":
                k = D.reduce(
                  (v, F) => v + (Number(F[y.uniqueName]) || 0),
                  0
                );
                break;
              case "avg":
                y?.formula && typeof y.formula == "function" && D.length > 0 ? k = D.reduce(
                  (v, F) => v + (y.formula?.(F) || 0),
                  0
                ) / D.length : k = D.reduce(
                  (v, F) => v + (Number(F[y.uniqueName]) || 0),
                  0
                ) / D.length;
                break;
              case "max":
                k = Math.max(
                  ...D.map(
                    (v) => Number(v[y.uniqueName]) || 0
                  )
                );
                break;
              case "min":
                k = Math.min(
                  ...D.map(
                    (v) => Number(v[y.uniqueName]) || 0
                  )
                );
                break;
              case "count":
                k = D.length;
                break;
              default:
                k = 0;
            }
          p.push(k);
        });
      }), p;
    });
    if (e.processedData && e.processedData.totals) {
      const f = ["Total"];
      u.forEach(() => {
        a.forEach((p) => {
          const w = e.processedData.totals[p.uniqueName] || 0;
          f.push(w);
        });
      }), h.push(f);
    }
    const g = [d, ...h], m = x.utils.aoa_to_sheet(g), C = x.utils.decode_range(m["!ref"] ?? "A1:A1"), S = [];
    for (let f = 0; f <= C.e.c; f++)
      S[f] = { wch: f === 0 ? 15 : 12 };
    m["!cols"] = S;
    for (let f = 1; f <= h.length; f++)
      for (let p = 1; p <= u.length * a.length; p++) {
        const w = x.utils.encode_cell({ r: f, c: p }), y = (p - 1) % a.length, D = a[y];
        if (D && D.format && m[w])
          if (D.format.type === "currency")
            m[w].z = D.format.currency === "USD" ? '"$"#,##0.00' : `"${D.format.currency}"#,##0.00`;
          else if (D.format.type === "number" && D.format.decimals !== void 0) {
            const k = "#,##0" + (D.format.decimals > 0 ? "." + "0".repeat(D.format.decimals) : "");
            m[w].z = k;
          } else D.format.type === "percentage" && (m[w].z = "0.00%", typeof m[w].v == "number" && (m[w].v = m[w].v / 100));
      }
    const b = x.utils.book_new();
    x.utils.book_append_sheet(b, m, "Pivot Table"), x.writeFile(b, `${t}.xlsx`);
  }
  /**
   * Opens a print dialog with formatted pivot table content
   * @param {PivotTableState<T>} state - The current state of the pivot table
   * @param {string} title - Optional title for the printed page
   */
  static openPrintDialog(e) {
    const t = E.convertToHtml(e), r = window.open("", "_blank");
    if (!r) {
      l.error("Failed to open print dialog");
      return;
    }
    r.document.write(t), r.document.close(), r.print();
  }
}
class et {
  constructor(e) {
    if (this.filterConfig = [], this.paginationConfig = {
      currentPage: 1,
      pageSize: 10,
      totalPages: 1
    }, this.autoAllColumnEnabled = !1, this.cache = /* @__PURE__ */ new Map(), this.listeners = /* @__PURE__ */ new Set(), !this.validateConfig(e))
      throw new Error("Invalid pivot table configuration");
    this.config = {
      ...e,
      defaultAggregation: e.defaultAggregation || "sum",
      isResponsive: e.isResponsive ?? !0
    }, this.state = this.initializeState(e), this.loadData();
  }
  setColumns(e) {
    throw new Error("Method not implemented.");
  }
  setRows(e) {
    throw new Error("Method not implemented.");
  }
  setData(e) {
    throw new Error("Method not implemented.");
  }
  validateConfig(e) {
    if (!e) return !1;
    if (e.dataSource) {
      const { type: t, url: r, file: s } = e.dataSource;
      if (t === "remote" && !r || t === "file" && !s) return !1;
    }
    return !0;
  }
  initializeState(e) {
    return {
      data: e.data || [],
      dataHandlingMode: "processed",
      rawData: e.data || [],
      processedData: { headers: [], rows: [], totals: {} },
      rows: e.rows || [],
      columns: e.columns || [],
      measures: this.normalizeMeasures(e.measures || []),
      sortConfig: [],
      rowSizes: this.initializeRowSizes(e.data || []),
      expandedRows: {},
      groupConfig: e.groupConfig || null,
      groups: [],
      selectedMeasures: this.normalizeMeasures(e.measures || []),
      selectedDimensions: e.dimensions || [],
      selectedAggregation: e.defaultAggregation || "sum",
      formatting: e.formatting || {},
      columnWidths: {},
      isResponsive: e.isResponsive ?? !0,
      rowGroups: [],
      columnGroups: [],
      filterConfig: [],
      paginationConfig: {
        currentPage: 1,
        pageSize: e.pageSize || 10,
        totalPages: 1
      }
    };
  }
  /**
   * Loads data from a file or URL.
   **/
  async loadData() {
    if (this.config.dataSource) {
      const { type: e, url: t, file: r } = this.config.dataSource;
      e === "remote" && t ? this.state.rawData = await this.fetchRemoteData(t) : e === "file" && r ? this.state.rawData = await this.readFileData(r) : l.error("Invalid data source configuration");
    } else this.config.data && (this.state.rawData = this.config.data);
    this.state.rowSizes = this.initializeRowSizes(this.state.rawData), this.ensureSyntheticAllColumn(), this.state.processedData = this.generateProcessedDataForDisplay(), this.state.groupConfig && this.applyGrouping();
  }
  // Ensure we always have a non-empty column axis and data augmented for '__all__' when needed
  ensureSyntheticAllColumn() {
    if (!this.autoAllColumnEnabled) return;
    const e = !this.state.columns || this.state.columns.length === 0, t = !!this.state.columns?.some(
      (r) => r.uniqueName === "__all__"
    );
    if (e && (this.state.columns = [
      { uniqueName: "__all__", caption: "All" }
    ]), e || t) {
      const r = (s) => {
        const a = Array.isArray(s) ? s : [];
        let i = !1;
        const n = a.map((o) => o && o.__all__ === void 0 ? (i = !0, { ...o, __all__: "All" }) : o);
        return i ? n : a;
      };
      this.config.data = r(this.config.data), this.state.rawData = r(this.state.rawData), this.state.data = r(this.state.data);
    }
  }
  setDataHandlingMode(e) {
    this.state.dataHandlingMode = e, this.refreshData(), this._emit();
  }
  getDataHandlingMode() {
    return this.state.dataHandlingMode;
  }
  // Allow external callers (e.g., ConnectService) to enable/disable synthetic column mode
  setAutoAllColumn(e) {
    this.autoAllColumnEnabled = e, e && (this.ensureSyntheticAllColumn(), this.state.processedData = this.generateProcessedDataForDisplay(), this._emit());
  }
  /**
   * Updates the engine's data source and clears filters
   * This method allows external components to update the data with fresh filtering state
   * @param {T[]} newData - The new data to use as the source
   * @param {boolean} clearFilters - Whether to clear existing filters (default: true)
   * @public
   */
  updateDataSource(e, t = !0) {
    this.config.data = [...e], this.state.data = [...e], this.state.rawData = [...e], t && (this.filterConfig = [], this.state.filterConfig = []), this.ensureSyntheticAllColumn(), this.refreshData(), this._emit();
  }
  /**
   * Loads data from a file or URL.
   * @param {File | string} source - The file or URL to load data from.
   * @public
   * @returns {Promise<void>} A promise that resolves when the data is loaded.
   **/
  async fetchRemoteData(e) {
    try {
      const t = await fetch(e);
      if (!t.ok)
        throw new Error(`Failed to fetch data from ${e}`);
      return await t.json();
    } catch (t) {
      return l.error("Error fetching remote data:", t), [];
    }
  }
  /**
   *  Process the data to be displayed in the table.
   * @param {T[]} data - The data to process.
   * @returns {ProcessedData} The processed data including headers, rows, and totals.
   * @private
   **/
  async readFileData(e) {
    return new Promise((t, r) => {
      const s = new FileReader();
      s.onload = (a) => {
        try {
          const i = JSON.parse(a.target?.result);
          t(i);
        } catch (i) {
          r(i);
        }
      }, s.onerror = (a) => r(a), s.readAsText(e);
    });
  }
  /**
   * Initializes row sizes for the pivot table.
   * @param {T[]} data - The data to initialize row sizes for.
   * @returns {RowSize[]} An array of row sizes.
   * @private
   */
  initializeRowSizes(e) {
    return e.map((t, r) => ({ index: r, height: 40 }));
  }
  /**
   * Processes the data for the pivot table.
   * @param {T[]} data - The data to process.
   * @returns {ProcessedData} The processed data including headers, rows, and totals.
   * @private
   */
  generateProcessedDataForDisplay() {
    let e = this.state.rawData;
    if (this.state.dataHandlingMode === "processed" && this.state.sortConfig.length > 0) {
      const t = this.state.sortConfig[0];
      this.state.groups.length > 0 ? e = this.state.groups.flatMap((r) => r.items) : e = this.sortData(this.state.rawData, t);
    }
    return {
      headers: this.generateHeaders(),
      rows: this.generateRows(e),
      totals: this.calculateTotals(e)
    };
  }
  /**
   * Generates headers for the pivot table.
   * @returns {string[]} An array of header strings.
   * @private
   */
  generateHeaders() {
    const e = this.state.dataHandlingMode === "raw", t = this.state.rows ? this.state.rows.map(
      (s) => e ? s.uniqueName : s.caption || s.uniqueName
    ) : [], r = this.state.columns ? this.state.columns.map(
      (s) => e ? s.uniqueName : s.caption || s.uniqueName
    ) : [];
    return [...t, ...r];
  }
  /**
   * Generates rows for the pivot table.
   * @param {T[]} data - The data to generate rows from.
   * @returns {CellValue[][]} A 2D array representing the rows.
   * @private
   */
  /**
   * Generates rows for the pivot table with enhanced formatting.
   * @param {T[]} data - The data to generate rows from.
   * @returns {CellValue[][]} A 2D array representing the rows.
   * @private
   */
  generateRows(e) {
    return !e || !this.state.rows || !this.state.columns ? [] : e.map((t) => [
      ...this.state.rows.map((r) => t[r.uniqueName]),
      ...this.state.columns.map((r) => t[r.uniqueName]),
      ...this.state.measures.map((r) => {
        const s = this.calculateMeasureValue(t, r);
        return this.formatValue(s, r.uniqueName);
      })
    ]);
  }
  /**
   * Calculates the value for a specific measure.
   * @param {T} item - The data item.
   * @param {MeasureConfig} measure - The measure configuration.
   * @returns {number} The calculated measure value.
   * @private
   */
  calculateMeasureValue(e, t) {
    return t.formula && typeof t.formula == "function" ? t.formula(e) : e[t.uniqueName] || 0;
  }
  /**
   * Calculates totals for each measure in the pivot table.
   * @param {T[]} data - The data to calculate totals from.
   * @returns {Record<string, number>} An object with measure names as keys and their totals as values.
   * @private
   */
  calculateTotals(e) {
    const t = {};
    return this.state.measures.forEach((r) => {
      const { uniqueName: s, aggregation: a } = r;
      let i = 0;
      a === "sum" ? i = e.reduce(
        (n, o) => n + (o[s] || 0),
        0
      ) : a === "avg" ? i = e.reduce(
        (n, o) => n + (o[s] || 0),
        0
      ) / e.length : a === "max" ? i = Math.max(
        ...e.map((n) => n[s] || 0)
      ) : a === "min" ? i = Math.min(
        ...e.map((n) => n[s] || 0)
      ) : a === "count" && (i = e.length), t[s] = i;
    }), t;
  }
  /**
   * Subscribe to state changes. Returns an unsubscribe function.
   */
  subscribe(e) {
    return this.listeners.add(e), e(this.getState()), () => this.listeners.delete(e);
  }
  /**
   * Emit state changes to all subscribers.
   * CRITICAL FIX: Wrap each listener in try-catch to prevent one subscriber error from crashing others
   */
  _emit() {
    const e = this.getState();
    this.listeners.forEach((t) => {
      try {
        t(e);
      } catch (r) {
        l.error("Error in pivot engine subscriber:", r);
      }
    });
  }
  /**
   * Sets the measures for the pivot table.
   * @param {MeasureConfig[]} measureFields - The measure configurations to set.
   * @public
   */
  setMeasures(e) {
    this.state.selectedMeasures = this.normalizeMeasures(e), this.state.processedData = this.generateProcessedDataForDisplay(), this.updateAggregates(), this._emit();
  }
  /**
   * Sets the dimensions for the pivot table.
   * @param {Dimension[]} dimensionFields - The dimension configurations to set.
   * @public
   */
  setDimensions(e) {
    this.state.selectedDimensions = e, this.state.processedData = this.generateProcessedDataForDisplay(), this.updateAggregates(), this.refreshData(), this._emit();
  }
  /**
   * Sets the aggregation type for the pivot table.
   * @param {AggregationType} type - The aggregation type to set.
   * @public
   */
  setAggregation(e) {
    this.state.selectedAggregation = e, this.state.processedData = this.generateProcessedDataForDisplay(), this.updateAggregates(), this.refreshData(), this._emit();
  }
  /**
   * Sets the row groups for the pivot table.
   * @param {Group[]} rowGroups - The row groups to set.
   * @public
   */
  setRowGroups(e) {
    this.state.rowGroups = e, this.state.processedData = this.generateProcessedDataForDisplay(), this.updateAggregates(), this._emit();
  }
  /**
   * Sets the column groups for the pivot table.
   * @param {Group[]} columnGroups - The column groups to set.
   * @public
   */
  setColumnGroups(e) {
    this.state.columnGroups = e, this.state.processedData = this.generateProcessedDataForDisplay(), this.updateAggregates(), this._emit();
  }
  /**
   * Enhanced formatValue method with comprehensive formatting support
   * @param {CellValue} value - The value to format.
   * @param {string} field - The field name to use for formatting.
   * @returns {string} The formatted value as a string.
   * @public
   */
  formatValue(e, t) {
    if (e == null || typeof e == "number" && isNaN(e)) {
      const a = this.getFieldFormat(t);
      return a && a.nullValue !== void 0 ? a.nullValue === null ? "" : String(a.nullValue) : "";
    }
    const r = this.getFieldFormat(t), s = this.state.measures.find((a) => a.uniqueName === t);
    if (!r) {
      if (s?.aggregation === "avg") {
        const a = Number(e);
        return Number.isFinite(a) ? String(Math.round(a)) : String(e);
      }
      return String(e);
    }
    try {
      const a = s?.aggregation === "avg" && typeof r.decimals != "number" ? { ...r, decimals: 0 } : r;
      return this.applyFormatting(e, a);
    } catch (a) {
      return l.error(`Error formatting value for field ${t}:`, a), String(e);
    }
  }
  /**
   * Get formatting configuration for a field
   * @param {string} field - The field name
   * @returns {FormatOptions | null} The format configuration
   * @private
   */
  getFieldFormat(e) {
    const t = this.state.measures.find((s) => s.uniqueName === e);
    if (t && t.format)
      return t.format;
    const r = this.state.formatting[e];
    return r || null;
  }
  /**
   * Apply comprehensive formatting to a value
   * @param {CellValue} value - The value to format
   * @param {FormatOptions} format - The format configuration
   * @returns {string} The formatted value
   * @private
   */
  applyFormatting(e, t) {
    let r = parseFloat(String(e));
    if (isNaN(r))
      return String(e);
    t.percent && (r = r * 100);
    const s = typeof t.decimals == "number" ? t.decimals : 2;
    let a;
    switch (t.type) {
      case "currency":
        a = this.formatCurrency(r, t, s);
        break;
      case "percentage":
        a = this.formatPercentage(r, t, s);
        break;
      case "date":
        a = this.formatDate(e, t);
        break;
      case "number":
      default:
        a = this.formatNumber(r, t, s);
        break;
    }
    return a = this.applyCustomSeparators(a, t), a;
  }
  /**
   * Format as currency
   * @param {number} num - The number to format
   * @param {FormatOptions} format - The format configuration
   * @param {number} decimals - Number of decimal places
   * @returns {string} The formatted currency value
   * @private
   */
  formatCurrency(e, t, r) {
    const s = t.currency || "USD", a = t.locale || "en-US";
    let i = new Intl.NumberFormat(a, {
      style: "currency",
      currency: s,
      minimumFractionDigits: r,
      maximumFractionDigits: r
    }).format(e);
    if (t.align === "right" || t.currencyAlign === "right") {
      const n = i.replace(/[\d.,\s]/g, "");
      i = i.replace(n, "").trim() + " " + n;
    } else t.align === "left" || t.currencyAlign;
    return i;
  }
  /**
   * Format as percentage
   * @param {number} num - The number to format
   * @param {FormatOptions} format - The format configuration
   * @param {number} decimals - Number of decimal places
   * @returns {string} The formatted percentage value
   * @private
   */
  formatPercentage(e, t, r) {
    const s = t.locale || "en-US";
    return new Intl.NumberFormat(s, {
      style: "percent",
      minimumFractionDigits: r,
      maximumFractionDigits: r
    }).format(e / 100);
  }
  /**
   * Format as number
   * @param {number} num - The number to format
   * @param {FormatOptions} format - The format configuration
   * @param {number} decimals - Number of decimal places
   * @returns {string} The formatted number value
   * @private
   */
  formatNumber(e, t, r) {
    const s = t.locale || "en-US";
    return new Intl.NumberFormat(s, {
      minimumFractionDigits: r,
      maximumFractionDigits: r
    }).format(e);
  }
  /**
   * Format as date
   * @param {CellValue} value - The date value to format
   * @param {FormatOptions} format - The format configuration
   * @returns {string} The formatted date value
   * @private
   */
  formatDate(e, t) {
    const r = t.locale || "en-US";
    try {
      return new Date(e).toLocaleDateString(r, {
        dateStyle: "medium"
      });
    } catch {
      return String(e);
    }
  }
  /**
   * Apply custom thousand and decimal separators
   * @param {string} formattedValue - The pre-formatted value
   * @param {FormatOptions} format - The format configuration
   * @returns {string} The value with custom separators applied
   * @private
   */
  applyCustomSeparators(e, t) {
    let r = e;
    if (t.decimalSeparator && t.decimalSeparator !== ".") {
      const s = r.lastIndexOf(".");
      s !== -1 && (r = r.substring(0, s) + t.decimalSeparator + r.substring(s + 1));
    }
    return t.thousandSeparator !== void 0 && (t.thousandSeparator === "" ? r = r.replace(/,/g, "") : t.thousandSeparator !== "," && (r = r.replace(/,/g, t.thousandSeparator))), r;
  }
  /**
   * Calculate aggregated value for a cell intersection
   * @param {string} rowValue - The row value
   * @param {string} columnValue - The column value
   * @param {MeasureConfig} measure - The measure configuration
   * @param {string} rowFieldName - The row field name
   * @param {string} columnFieldName - The column field name
   * @returns {number} The calculated aggregated value
   * @public
   */
  calculateCellValue(e, t, r, s, a) {
    const n = (this.config.data || this.state.rawData).filter(
      (u) => u[s] === e && u[a] === t
    );
    if (n.length === 0)
      return 0;
    let o = 0;
    switch (r.aggregation) {
      case "sum":
        o = n.reduce(
          (u, d) => u + (d[r.uniqueName] || 0),
          0
        );
        break;
      case "avg":
        o = n.reduce(
          (u, d) => u + (d[r.uniqueName] || 0),
          0
        ) / n.length;
        break;
      case "max":
        o = Math.max(
          ...n.map((u) => u[r.uniqueName] || 0)
        );
        break;
      case "min":
        o = Math.min(
          ...n.map((u) => u[r.uniqueName] || 0)
        );
        break;
      case "count":
        o = n.length;
        break;
      default:
        o = 0;
    }
    return o;
  }
  /**
   * Get text alignment for a field
   * @param {string} field - The field name
   * @returns {string} The text alignment ('left', 'right', 'center')
   * @public
   */
  getFieldAlignment(e) {
    const t = `alignment:${e}`;
    if (this.cache.has(t))
      return this.cache.get(t);
    const r = this.getFieldFormat(e);
    let s;
    return r && r.align ? s = r.align : r && r.type === "currency" && r.currencyAlign ? s = r.currencyAlign : s = this.state.measures.find((i) => i.uniqueName === e) ? "right" : "left", this.cache.set(t, s), s;
  }
  /**
   * Update formatting configuration for a specific field
   * @param {string} field - The field name
   * @param {FormatOptions} format - The format configuration
   * @public
   */
  updateFieldFormatting(e, t) {
    const r = this.state.measures.find((a) => a.uniqueName === e);
    r && (r.format = t), this.state.formatting[e] = t;
    const s = `alignment:${e}`;
    this.cache.delete(s), this.state.processedData = this.generateProcessedDataForDisplay(), this._emit();
  }
  /**
   * Sorts the pivot table data based on the specified field and direction.
   * @param {string} field - The field to sort by.
   * @param {'asc' | 'desc'} direction - The sort direction.
   * @public
   */
  sort(e, t) {
    const r = this.state.measures.find((a) => a.uniqueName === e), s = {
      field: e,
      direction: t,
      type: r ? "measure" : "dimension",
      aggregation: r?.aggregation
    };
    this.state.sortConfig = [s], this.applySort();
  }
  applySort() {
    if (this.state.dataHandlingMode === "raw") {
      const e = this.sortData(
        this.state.rawData,
        this.state.sortConfig[0]
      );
      this.state.data = e, this.state.rawData = e, this.state.processedData = this.generateProcessedDataForDisplay(), this.updateAggregates();
    } else if (this.state.groups.length > 0)
      this.state.groups = this.sortGroups(
        this.state.groups,
        this.state.sortConfig[0]
      ), this.state.processedData = this.generateProcessedDataForDisplay(), this.updateAggregates();
    else {
      const e = this.sortData(
        this.state.rawData,
        this.state.sortConfig[0]
      );
      this.state.data = e, this.state.rawData = e, this.state.processedData = this.generateProcessedDataForDisplay(), this.updateAggregates();
    }
    this._emit();
  }
  sortData(e, t) {
    return [...e].sort((r, s) => {
      let a = this.getFieldValue(r, t), i = this.getFieldValue(s, t);
      return typeof a == "string" && (a = a.toLowerCase()), typeof i == "string" && (i = i.toLowerCase()), a < i ? t.direction === "asc" ? -1 : 1 : a > i ? t.direction === "asc" ? 1 : -1 : 0;
    });
  }
  getFieldValue(e, t) {
    if (t.type === "measure") {
      const r = this.state.measures.find(
        (s) => s.uniqueName === t.field
      );
      if (r && r.formula)
        return r.formula(e);
    }
    return e[t.field];
  }
  sortGroups(e, t) {
    return [...e].sort((r, s) => {
      let a, i;
      if (t.type === "measure")
        a = r.aggregates[`${t.aggregation}_${t.field}`] || 0, i = s.aggregates[`${t.aggregation}_${t.field}`] || 0;
      else {
        const n = r.key ? r.key.split("|") : [], o = s.key ? s.key.split("|") : [], u = this.state.rows?.[0]?.uniqueName, d = this.state.columns?.[0]?.uniqueName;
        t.field === u ? (a = n[0] || "", i = o[0] || "") : t.field === d ? (a = n[1] || "", i = o[1] || "") : (a = r.items[0]?.[t.field] || "", i = s.items[0]?.[t.field] || ""), typeof a == "string" && (a = a.toLowerCase()), typeof i == "string" && (i = i.toLowerCase());
      }
      return a < i ? t.direction === "asc" ? -1 : 1 : a > i ? t.direction === "asc" ? 1 : -1 : 0;
    });
  }
  /**
   * Updates aggregates for all groups in the pivot table.
   * @private
   */
  updateAggregates() {
    const e = (t) => {
      this.state.measures.forEach((r) => {
        const s = `${this.state.selectedAggregation}_${r.uniqueName}`;
        if (r.formula && typeof r.formula == "function") {
          const a = t.items.map(
            (i) => r.formula ? r.formula(i) : 0
          );
          t.aggregates[s] = re(
            a.map((i) => ({ value: i })),
            "value",
            r.aggregation || this.state.selectedAggregation
          );
        } else
          t.aggregates[s] = re(
            t.items,
            r.uniqueName,
            r.aggregation || this.state.selectedAggregation
          );
      }), t.subgroups && t.subgroups.forEach(e);
    };
    this.state.groups.forEach(e);
  }
  /**
   * Applies grouping to the pivot table data.
   * @private
   */
  applyGrouping(e) {
    if (!this.state.groupConfig) return;
    const { rowFields: t, columnFields: r, grouper: s } = this.state.groupConfig;
    if (!t || !r || !s) {
      l.error("Invalid groupConfig:", this.state.groupConfig);
      return;
    }
    const a = e || this.config.data || [], i = {
      ...this.config,
      data: a
    }, { rawData: n, groups: o } = be(
      i,
      this.state.sortConfig[0] || null,
      this.state.groupConfig
    );
    this.state.rawData = n, this.state.groups = o, this.updateAggregates(), this.state.processedData = this.generateProcessedDataForDisplay();
  }
  /**
   * Sets the group configuration for the pivot table.
   * @param {GroupConfig | null} groupConfig - The group configuration to set.
   * @public
   */
  setGroupConfig(e) {
    this.state.groupConfig = e, e ? this.applyGrouping() : (this.state.groups = [], this.state.processedData = this.generateProcessedDataForDisplay()), this._emit();
  }
  /**
   * Returns the grouped data.
   * @returns {Group[]} An array of grouped data.
   * @public
   */
  getGroupedData() {
    return this.state.groups;
  }
  /**
   * Returns the current state of the pivot table.
   * @returns {PivotTableState<T>} The current state of the pivot table.
   * @public
   */
  getState() {
    return { ...this.state };
  }
  /**
   * Resets the pivot table to its initial state.
   * @public
   */
  reset() {
    this.state = {
      ...this.state,
      rawData: this.config.data || [],
      processedData: this.generateProcessedDataForDisplay(),
      sortConfig: [],
      rowSizes: this.initializeRowSizes(this.config.data || []),
      expandedRows: {},
      groupConfig: this.config.groupConfig || null,
      groups: []
    }, this.state.groupConfig && this.applyGrouping(), this._emit();
  }
  /**
   * Resizes a specific row in the pivot table.
   * @param {number} index - The index of the row to resize.
   * @param {number} height - The new height for the row.
   * @public
   */
  resizeRow(e, t) {
    const r = this.state.rowSizes.findIndex((s) => s.index === e);
    r !== -1 && (this.state.rowSizes[r].height = Math.max(20, t), this._emit());
  }
  /**
   * Toggles the expansion state of a row.
   * @param {string} rowId - The ID of the row to toggle.
   * @public
   */
  toggleRowExpansion(e) {
    this.state.expandedRows[e] = !this.state.expandedRows[e], this._emit();
  }
  /**
   * Checks if a row is expanded.
   * @param {string} rowId - The ID of the row to check.
   * @returns {boolean} True if the row is expanded, false otherwise.
   * @public
   */
  isRowExpanded(e) {
    return !!this.state.expandedRows[e];
  }
  /**
   * Handles dragging a row to a new position.
   * This method now correctly operates on state.rowGroups.
   * @param {number} fromIndex - The original index of the row.
   * @param {number} toIndex - The new index for the row.
   * @public
   */
  dragRow(e, t) {
    if (!this.validateDragOperation(
      e,
      t,
      this.state.rowGroups.length
    ))
      return;
    const r = [...this.state.rowGroups], [s] = r.splice(e, 1);
    r.splice(t, 0, s), this.state.rowGroups = r, typeof this.config.onRowDragEnd == "function" && this.config.onRowDragEnd(e, t, this.state.rowGroups), this._emit();
  }
  /**
   * Handles dragging a column to a new position.
   * This method now correctly operates on state.columnGroups.
   * @param {number} fromIndex - The original index of the column.
   * @param {number} toIndex - The new index for the column.
   * @public
   */
  dragColumn(e, t) {
    if (this.validateDragOperation(
      e,
      t,
      this.state.columnGroups.length
    ))
      try {
        const r = [...this.state.columnGroups], [s] = r.splice(e, 1);
        if (r.splice(t, 0, s), this.state.columnGroups = r, typeof this.config.onColumnDragEnd == "function") {
          const a = r.map((i) => ({
            uniqueName: i.uniqueName ?? i.key ?? "",
            caption: i.caption ?? i.key ?? ""
          }));
          this.config.onColumnDragEnd(e, t, a);
        }
        this._emit();
      } catch (r) {
        l.error("Error during column drag operation:", r);
      }
  }
  // Ensure this validation method also prevents dragging to the same spot
  validateDragOperation(e, t, r) {
    if (e === t)
      return !1;
    const s = e >= 0 && t >= 0 && e < r && t < r;
    return s || l.warn(
      `Invalid drag indices: from ${e} to ${t} with length ${r}`
    ), s;
  }
  /**
   * Applies filters to the data
   * @param {FilterConfig[]} filters - Array of filter configurations
   * @public
   */
  applyFilters(e) {
    this.filterConfig = e, this.refreshData(), this._emit();
  }
  /**
   * Sets pagination configuration
   * @param {PaginationConfig} config - Pagination configuration
   * @public
   */
  setPagination(e) {
    this.paginationConfig = {
      ...this.paginationConfig,
      ...e
    }, this.refreshData(), this._emit();
  }
  /**
   * Returns the current pagination configuration
   * @returns {PaginationConfig}
   * @public
   */
  getPagination() {
    return this.paginationConfig;
  }
  /**
   * Refreshes data with current filters and pagination
   * @private
   */
  refreshData() {
    const e = this.getDataForCurrentMode();
    let t;
    this.state.dataHandlingMode === "processed" && this.hasAggregatedFilters() ? t = this.filterProcessedData(e) : t = this.filterData(e), this.paginationConfig.totalPages = Math.ceil(
      t.length / this.paginationConfig.pageSize
    ), t = this.paginateData(t), this.state.dataHandlingMode === "raw" ? (this.state.data = t, this.state.rawData = t, this.state.processedData = this.generateProcessedDataForDisplay()) : (this.state.data = t, this.state.rawData = t, this.state.groupConfig && this.applyGrouping(t), this.state.processedData = this.generateProcessedDataForDisplay());
  }
  /**
   * Gets the appropriate data source based on the current data handling mode
   * @private
   */
  getDataForCurrentMode() {
    return [...this.config.data || []];
  }
  /**
   * Filters data based on filter configuration
   * @param {T[]} data - Data to filter
   * @private
   */
  filterData(e) {
    return this.filterConfig.length ? e.filter(
      (t) => this.filterConfig.every((r) => {
        const s = t[r.field], a = typeof s == "number" ? Number(r.value) : r.value;
        switch (r.operator) {
          case "equals":
            return typeof s == "string" && typeof a == "string" ? s.toLowerCase() === a.toLowerCase() : s === a;
          case "contains":
            return String(s).toLowerCase().includes(String(a).toLowerCase());
          case "greaterThan":
            return Number(s) > Number(a);
          case "lessThan":
            return Number(s) < Number(a);
          case "between": {
            const i = a;
            return s >= i[0] && s <= i[1];
          }
          default:
            return !0;
        }
      })
    ) : e;
  }
  /**
   * Paginates data based on pagination configuration
   * @param {T[]} data - Data to paginate
   * @private
   */
  paginateData(e) {
    const { currentPage: t, pageSize: r } = this.paginationConfig, s = (t - 1) * r;
    return e.slice(s, s + r);
  }
  /**
   * Gets current pagination state
   * @returns {PaginationConfig} Current pagination configuration
   * @public
   */
  getPaginationState() {
    return { ...this.paginationConfig };
  }
  /**
   * Gets current filter state
   * @returns {FilterConfig[]} Current filter configuration
   * @public
   */
  getFilterState() {
    return [...this.filterConfig];
  }
  /**
   * Exports the pivot table data to HTML and downloads the file.
   * @param {string} fileName - The name of the downloaded file (without extension).
   * @public
   */
  exportToHTML(e = "pivot-table") {
    l.info("PivotEngine.exportToHTML called with fileName:", e), l.info(
      "PivotEngine state rawData length:",
      this.state.rawData?.length || 0
    ), E.exportToHTML(this.getState(), e);
  }
  /**
   * Exports the pivot table data to PDF and downloads the file.
   * @param {string} fileName - The name of the downloaded file (without extension).
   * @public
   */
  exportToPDF(e = "pivot-table") {
    l.info("PivotEngine.exportToPDF called with fileName:", e), l.info(
      "PivotEngine state rawData length:",
      this.state.rawData?.length || 0
    ), E.exportToPDF(this.getState(), e);
  }
  /**
   * Exports the pivot table data to Excel and downloads the file.
   * @param {string} fileName - The name of the downloaded file (without extension).
   * @public
   */
  exportToExcel(e = "pivot-table") {
    l.info("PivotEngine.exportToExcel called with fileName:", e), l.info(
      "PivotEngine state rawData length:",
      this.state.rawData?.length || 0
    ), E.exportToExcel(this.getState(), e);
  }
  /**
   * Opens a print dialog with the formatted pivot table.
   * @public
   */
  openPrintDialog() {
    E.openPrintDialog(this.getState());
  }
  // Add these methods to your PivotEngine class to fix drag functionality
  /**
   * Handles dragging a data row (product) to a new position
   * This method operates on the actual data items, not groups
   * @param {number} fromIndex - The original index of the product in unique products list
   * @param {number} toIndex - The new index for the product in unique products list
   * @public
   */
  dragDataRow(e, t) {
    const r = [
      ...new Set(this.state.data.map((s) => s.product))
    ].filter((s) => typeof s == "string");
    if (this.validateDragOperation(e, t, r.length))
      try {
        const s = r[e], a = r[t];
        l.info(`Reordering products: ${s} -> ${a}`);
        const i = [...this.state.data], n = [...r], [o] = n.splice(e, 1);
        n.splice(t, 0, o), i.sort((u, d) => {
          const h = n.indexOf(u.product), g = n.indexOf(d.product);
          return h - g;
        }), this.state.data = i, this.state.rawData = i, this.state.processedData = this.generateProcessedDataForDisplay(), this.state.groups.length > 0 && this.updateAggregates(), typeof this.config.onRowDragEnd == "function" && this.config.onRowDragEnd(e, t, this.state.rowGroups);
      } catch (s) {
        l.error("Error during data row drag operation:", s);
      }
  }
  /**
   * Handles dragging a data column (region) to a new position
   * This method operates on the actual data structure, not groups
   * @param {number} fromIndex - The original index of the region
   * @param {number} toIndex - The new index for the region
   * @public
   */
  dragDataColumn(e, t) {
    const r = [
      ...new Set(this.state.data.map((s) => s.region))
    ].filter((s) => typeof s == "string");
    if (this.validateDragOperation(e, t, r.length))
      try {
        const s = [...this.state.data], a = [...r], [i] = a.splice(e, 1);
        if (a.splice(t, 0, i), this.state.columns && this.state.columns.length > 0) {
          const n = [...this.state.columns];
          n.sort((o, u) => {
            const d = a.indexOf(o.uniqueName), h = a.indexOf(u.uniqueName);
            return d === -1 ? 1 : h === -1 ? -1 : d - h;
          }), this.state.columns = n;
        }
        if (this.state.data = s, this.state.rawData = s, this.state.processedData = this.generateProcessedDataForDisplay(), this.state.groups.length > 0 && this.updateAggregates(), typeof this.config.onColumnDragEnd == "function") {
          const n = a.map((o) => ({
            uniqueName: o,
            caption: o
          }));
          this.config.onColumnDragEnd(e, t, n);
        }
      } catch (s) {
        l.error("Error during data column drag operation:", s);
      }
  }
  /**
   * Alternative method: Reorder products by their names directly
   * This is more direct for your UI implementation
   * @param {string} fromProduct - Name of the product being moved
   * @param {string} toProduct - Name of the product to move before/after
   * @param {'before' | 'after'} position - Whether to place before or after target
   * @public
   */
  reorderProductsByName(e, t, r = "before") {
    try {
      const s = [
        ...new Set(this.state.data.map((o) => o.product))
      ], a = s.indexOf(e), i = s.indexOf(t);
      if (a === -1 || i === -1) {
        l.warn("Invalid product names for reordering:", {
          fromProduct: e,
          toProduct: t
        });
        return;
      }
      const n = r === "after" ? i + 1 : i;
      this.dragDataRow(a, n);
    } catch (s) {
      l.error("Error reordering products by name:", s);
    }
  }
  //swap logic
  // Also add a method to get the custom region order:
  getCustomRegionOrder() {
    return this.state.customRegionOrder ?? null;
  }
  // GENERIC CORE ENGINE METHODS - Works with any field names
  /**
   * Generic method to swap data rows based on the configured row field
   * Works with any field name (product, country, customer, etc.)
   */
  swapDataRows(e, t) {
    const r = this.state.rows && this.state.rows.length > 0 ? this.state.rows[0] : null;
    if (!r) {
      l.warn("No row field configured for swapping");
      return;
    }
    const s = r.uniqueName, a = this.state.customRowOrder;
    let i;
    if (a && a.fieldName === s && a.order ? i = [...a.order] : i = [
      ...new Set((this.config.data || []).map((n) => n[s]))
    ].filter(
      (n) => typeof n == "string" && n !== null && n !== void 0
    ), l.info(
      `Row swap validation: ${s} has ${i.length} unique values:`,
      i
    ), e < 0 || t < 0 || e >= i.length || t >= i.length) {
      l.warn("Invalid indices for row swap operation:", {
        fromIndex: e,
        toIndex: t,
        totalRows: i.length,
        fieldName: s
      });
      return;
    }
    if (e !== t)
      try {
        const n = i[e], o = i[t], u = [...this.config.data || []], d = [...i];
        d[e] = o, d[t] = n, this.state.customRowOrder = {
          fieldName: s,
          order: d
        }, u.sort((h, g) => {
          const m = d.indexOf(h[s]), C = d.indexOf(g[s]);
          return m - C;
        }), this.config.data = u, this.state.rawData = u, this.state.data = u, this.state.processedData = this.generateProcessedDataForDisplay(), this.state.groups.length > 0 && this.updateAggregates(), typeof this.config.onRowDragEnd == "function" && this.config.onRowDragEnd(e, t, this.state.rowGroups), l.info(`Row swap completed successfully for field: ${s}`), this._emit();
      } catch (n) {
        l.error("Error during row swap operation:", n);
      }
  }
  /**
   * Generic method to swap data columns based on the configured column field
   * Works with any field name (region, category, department, etc.)
   */
  swapDataColumns(e, t) {
    const r = this.state.columns && this.state.columns.length > 0 ? this.state.columns[0] : null;
    if (!r) {
      l.warn("No column field configured for swapping");
      return;
    }
    const s = r.uniqueName, a = this.state.customColumnOrder;
    let i;
    if (a && a.fieldName === s && a.order ? i = [...a.order] : i = [
      ...new Set((this.config.data || []).map((n) => n[s]))
    ].filter(
      (n) => typeof n == "string" && n !== null && n !== void 0
    ), l.info(
      `Column swap validation: ${s} has ${i.length} unique values:`,
      i
    ), e < 0 || t < 0 || e >= i.length || t >= i.length) {
      l.warn("Invalid indices for column swap operation:", {
        fromIndex: e,
        toIndex: t,
        totalColumns: i.length,
        fieldName: s
      });
      return;
    }
    if (e !== t)
      try {
        const n = i[e], o = i[t], u = [...i];
        if (u[e] = o, u[t] = n, this.state.columns && this.state.columns.length > 0) {
          const d = [...this.state.columns], h = d.findIndex(
            (m) => m.uniqueName === n
          ), g = d.findIndex(
            (m) => m.uniqueName === o
          );
          h !== -1 && g !== -1 && ([d[h], d[g]] = [
            d[g],
            d[h]
          ], this.state.columns = d, l.info("Updated column configurations"));
        }
        if (this.state.customColumnOrder = {
          fieldName: s,
          order: u
        }, this.state.processedData = this.generateProcessedDataForDisplay(), this.state.groups.length > 0 && this.updateAggregates(), typeof this.config.onColumnDragEnd == "function") {
          const d = u.map((h) => ({
            uniqueName: h,
            caption: h
          }));
          this.config.onColumnDragEnd(e, t, d);
        }
        l.info(
          `Column swap completed successfully for field: ${s}`
        ), this._emit();
      } catch (n) {
        l.error("Error during column swap operation:", n);
      }
  }
  /**
   * Generic method to get unique values for any field
   * Uses the original config data, not the paginated state data
   * Utility method for UI components
   */
  getUniqueFieldValues(e) {
    const t = this.config.data || [];
    l.info(
      `Getting unique values for field: ${e} from ${t.length} total items`
    );
    const r = [
      ...new Set(t.map((s) => s[e]))
    ].filter(
      (s) => typeof s == "string" && s !== null && s !== void 0
    );
    return l.info(
      `Found ${r.length} unique values for ${e}:`,
      r
    ), r;
  }
  // Also add this method to get unique values in their current custom order:
  /**
   * Get unique field values respecting any custom order that has been set
   */
  getOrderedUniqueFieldValues(e, t = !1) {
    const r = t ? this.state.customRowOrder : this.state.customColumnOrder;
    return r && r.fieldName === e && r.order ? (l.info(`Using custom order for ${e}:`, r.order), r.order) : this.getUniqueFieldValues(e);
  }
  /**
   * Generic method to get the configured row field name
   */
  getRowFieldName() {
    return this.state.rows && this.state.rows.length > 0 ? this.state.rows[0].uniqueName : null;
  }
  /**
   * Generic method to get the configured column field name
   */
  getColumnFieldName() {
    return this.state.columns && this.state.columns.length > 0 ? this.state.columns[0].uniqueName : null;
  }
  /**
   * Generic method to set custom field order
   * Can be used by UI to store custom arrangements
   */
  setCustomFieldOrder(e, t, r = !0) {
    const s = { fieldName: e, order: t };
    r ? this.state.customRowOrder = s : this.state.customColumnOrder = s, this.state.processedData = this.generateProcessedDataForDisplay(), this._emit();
  }
  /**
   * Get ordered column values if custom order exists
   */
  getOrderedColumnValues() {
    const e = this.state.customColumnOrder;
    return e && e.order && e.order.length > 0 ? (l.info(
      "Engine returning custom column order:",
      e.order
    ), e.order) : (l.info("Engine has no custom column order"), null);
  }
  /**
   * Get ordered row values if custom order exists
   */
  getOrderedRowValues() {
    const e = this.state.customRowOrder;
    return e && e.order && e.order.length > 0 ? (l.info("Engine returning custom row order:", e.order), e.order) : (l.info("Engine has no custom row order"), null);
  }
  /**
   * Method to swap raw data rows by index
   * This works directly with the raw data array regardless of pivot configuration
   */
  swapRawDataRows(e, t) {
    if (!this.state.data || this.state.data.length === 0) {
      l.warn("No data available for raw row swap");
      return;
    }
    if (e < 0 || t < 0 || e >= this.state.data.length || t >= this.state.data.length) {
      l.warn("Invalid indices for raw row swap operation:", {
        fromIndex: e,
        toIndex: t,
        totalRows: this.state.data.length
      });
      return;
    }
    if (e !== t)
      try {
        const r = [...this.state.data], s = r[e];
        r[e] = r[t], r[t] = s, this.state.data = r, this.state.rawData = r, this.state.dataHandlingMode === "raw" ? this.state.processedData = this.generateProcessedDataForDisplay() : this.state.processedData = this.generateProcessedDataForDisplay(), this._emit();
      } catch (r) {
        l.error("Error during raw row swap operation:", r);
      }
  }
  /**
   * Checks if any of the current filters are for aggregated measures
   * @private
   */
  hasAggregatedFilters() {
    return this.filterConfig.some((e) => e.field.includes("_") && this.state.measures.some(
      (t) => e.field === `${t.aggregation}_${t.uniqueName}`
    ));
  }
  /**
   * Filters processed data based on aggregated values
   * @private
   */
  filterProcessedData(e) {
    this.state.groupConfig && this.applyGrouping(e);
    const t = this.filterConfig.filter(
      (a) => a.field.includes("_") && this.state.measures.some(
        (i) => a.field === `${i.aggregation}_${i.uniqueName}`
      )
    ), r = this.filterConfig.filter(
      (a) => !a.field.includes("_") || !this.state.measures.some(
        (i) => a.field === `${i.aggregation}_${i.uniqueName}`
      )
    );
    let s = e;
    return r.length > 0 && (s = s.filter(
      (a) => r.every((i) => {
        const n = a[i.field], o = typeof n == "number" ? Number(i.value) : i.value;
        switch (i.operator) {
          case "equals":
            return typeof n == "string" && typeof o == "string" ? n.toLowerCase() === o.toLowerCase() : n === o;
          case "contains":
            return String(n).toLowerCase().includes(String(o).toLowerCase());
          case "greaterThan":
            return Number(n) > Number(o);
          case "lessThan":
            return Number(n) < Number(o);
          case "between": {
            const u = o;
            return n >= u[0] && n <= u[1];
          }
          default:
            return !0;
        }
      })
    )), t.length > 0 && (this.state.groupConfig && this.applyGrouping(s), s = this.state.groups.filter(
      (i) => t.every((n) => {
        const o = i.aggregates[n.field], u = typeof n.value == "string" ? Number(n.value) : n.value;
        switch (n.operator) {
          case "equals":
            return o === u;
          case "greaterThan":
            return o > u;
          case "lessThan":
            return o < u;
          case "between": {
            const d = Array.isArray(u) ? u : [u, u];
            return o >= Number(d[0]) && o <= Number(d[1]);
          }
          default:
            return !0;
        }
      })
    ).flatMap((i) => i.items)), s;
  }
  /**
   * Sets the layout for the pivot table (rows, columns, measures) in one call.
   * @param {AxisConfig[]} rows - Row fields
   * @param {AxisConfig[]} columns - Column fields
   * @param {MeasureConfig[]} measures - Measure fields
   * @public
   */
  setLayout(e, t, r) {
    this.state.rows = e || [], this.state.columns = t || [], this.state.measures = this.normalizeMeasures(r || []), this.state.selectedMeasures = [...this.state.measures], this.state.selectedDimensions = [
      ...e.map(
        (s) => ({
          field: s.uniqueName,
          label: s.caption || s.uniqueName,
          type: "string"
        })
      ),
      ...t.map(
        (s) => ({
          field: s.uniqueName,
          label: s.caption || s.uniqueName,
          type: "string"
        })
      )
    ], this.ensureSyntheticAllColumn(), this.state.processedData = this.generateProcessedDataForDisplay(), this.updateAggregates(), this._emit();
  }
  /**
   * Set only rows and columns, keeping measures unchanged
   * @param {AxisConfig[]} rows
   * @param {AxisConfig[]} columns
   * @public
   */
  setRowsAndColumns(e, t) {
    this.state.rows = e || [], this.state.columns = t || [], this.ensureSyntheticAllColumn(), this.state.processedData = this.generateProcessedDataForDisplay(), this.updateAggregates(), this._emit();
  }
  normalizeMeasures(e = []) {
    return (e || []).map((t) => !(!!t.caption && t.caption.trim().toLowerCase() !== t.uniqueName.toLowerCase()) && (t.aggregation || this.config.defaultAggregation) === "sum" ? { ...t, caption: `sum of ${t.uniqueName}` } : t);
  }
}
class tt {
  /** Returns all fields inferred from the engine's raw data */
  static getAvailableFields(e) {
    const r = e.getState().rawData || [];
    if (!Array.isArray(r) || r.length === 0) return [];
    const s = r.slice(0, Math.min(200, r.length)), a = /* @__PURE__ */ new Set();
    s.forEach(
      (n) => Object.keys(n || {}).forEach((o) => a.add(o))
    );
    const i = [];
    return a.forEach((n) => {
      i.push({ name: n, type: this.inferFieldType(s, n) });
    }), i.sort((n, o) => n.name.localeCompare(o.name));
  }
  /** Supported aggregations */
  static getSupportedAggregations() {
    return ["sum", "avg", "min", "max", "count"];
  }
  /** Change aggregation for a specific measure and refresh engine */
  static setMeasureAggregation(e, t, r) {
    const a = [...e.getState().measures], i = a.findIndex((n) => n.uniqueName === t);
    i === -1 ? a.push(this.buildMeasure(t, r)) : a[i] = { ...a[i], aggregation: r }, e.setMeasures(a);
  }
  /** Build complete layout (rows, columns, measures) from selection */
  static buildLayout(e) {
    const t = (e.rows || []).map((a) => ({
      uniqueName: a,
      caption: this.toCaption(a)
    })), r = (e.columns || []).map((a) => ({
      uniqueName: a,
      caption: this.toCaption(a)
    })), s = (e.values || []).map(
      (a) => this.buildMeasure(a.field, a.aggregation)
    );
    return { rows: t, columns: r, measures: s };
  }
  /** Build a single MeasureConfig */
  static buildMeasure(e, t) {
    const r = t || "sum";
    return {
      uniqueName: e,
      caption: `${this.capitalize(r)} of ${this.toCaption(e)}`,
      aggregation: r
    };
  }
  // --- helpers ---
  static inferFieldType(e, t) {
    let r = 0, s = 0;
    return e.forEach((a) => {
      const i = a?.[t];
      if (i == null) return;
      typeof i == "number" || !isNaN(Number(i)) && i !== "" ? r++ : this.looksLikeDate(i) && s++;
    }), r > 0 && r >= s ? "number" : s > 0 ? "date" : "string";
  }
  static looksLikeDate(e) {
    if (typeof e != "string" && !(e instanceof Date)) return !1;
    const t = Date.parse(String(e));
    return !isNaN(t);
  }
  static toCaption(e) {
    return e.replace(/([a-z0-9])([A-Z])/g, "$1 $2").replace(/[_-]+/g, " ").replace(/^\w/, (t) => t.toUpperCase());
  }
  static capitalize(e) {
    return e.charAt(0).toUpperCase() + e.slice(1);
  }
}
class We {
  constructor(e, t) {
    this.workers = [], this.taskQueue = [], this.nextTaskId = 0, this.workerUrl = e, this.workerCount = t || Math.max(1, navigator.hardwareConcurrency - 1), this.initializeWorkers();
  }
  /**
   * Initialize worker pool
   */
  initializeWorkers() {
    for (let e = 0; e < this.workerCount; e++) {
      const t = new Worker(this.workerUrl, { type: "module" });
      this.workers.push({
        worker: t,
        busy: !1
      });
    }
  }
  /**
   * Execute a task on an available worker
   */
  execute(e) {
    return new Promise((t, r) => {
      const a = {
        id: this.nextTaskId++,
        data: e,
        resolve: t,
        reject: r
      };
      this.taskQueue.push(a), this.processQueue();
    });
  }
  /**
   * Process the task queue
   */
  processQueue() {
    const e = this.workers.find((a) => !a.busy);
    if (!e || this.taskQueue.length === 0)
      return;
    const t = this.taskQueue.shift();
    if (!t) return;
    e.busy = !0, e.currentTaskId = t.id;
    const r = (a) => {
      const i = a.data;
      i.type === "CHUNK_PARSED" ? (e.worker.removeEventListener("message", r), e.worker.removeEventListener("error", s), e.busy = !1, e.currentTaskId = void 0, i.error ? t.reject(new Error(i.error)) : t.resolve(i), this.processQueue()) : i.type;
    }, s = (a) => {
      e.worker.removeEventListener("message", r), e.worker.removeEventListener("error", s), e.busy = !1, e.currentTaskId = void 0, t.reject(a), this.processQueue();
    };
    e.worker.addEventListener("message", r), e.worker.addEventListener("error", s), e.worker.postMessage(t.data);
  }
  /**
   * Terminate all workers
   */
  terminate() {
    this.workers.forEach(({ worker: e }) => e.terminate()), this.workers = [], this.taskQueue = [];
  }
  /**
   * Get number of active workers
   */
  getWorkerCount() {
    return this.workers.length;
  }
  /**
   * Get number of busy workers
   */
  getBusyCount() {
    return this.workers.filter((e) => e.busy).length;
  }
  /**
   * Get number of queued tasks
   */
  getQueuedCount() {
    return this.taskQueue.length;
  }
}
const X = class X {
  // 1 MB
  /**
   * Read file in chunks using Streams API
   */
  static async readFileInChunks(e, t = {}) {
    const r = t.chunkSizeBytes || this.DEFAULT_CHUNK_SIZE, s = e.size;
    let a = 0, i = 0;
    if (!e.stream)
      return this.readFileInChunksLegacy(e, t);
    try {
      const o = e.stream().getReader(), u = new TextDecoder(t.encoding || "utf-8");
      let d = "", h = !1, g = !1;
      for (; !h && !g; ) {
        if (t.signal?.aborted) {
          g = !0, o.cancel();
          break;
        }
        const { value: m, done: C } = await o.read();
        if (h = C, m)
          for (a += m.length, d += u.decode(m, { stream: !h }); d.length >= r || h && d.length > 0; ) {
            if (t.signal?.aborted) {
              g = !0, h || o.cancel();
              break;
            }
            const S = d.slice(0, r);
            d = d.slice(r);
            const b = Math.min(100, a / s * 100), f = {
              chunkId: i++,
              text: S,
              isFirstChunk: i === 1,
              isLastChunk: h && d.length === 0,
              progress: b
            };
            if (t.onChunk && await t.onChunk(f), t.onProgress && t.onProgress(b), h && d.length === 0)
              break;
          }
      }
      if (d.length > 0) {
        const m = {
          chunkId: i++,
          text: d,
          isFirstChunk: i === 1,
          isLastChunk: !0,
          progress: 100
        };
        t.onChunk && await t.onChunk(m), t.onProgress && t.onProgress(100);
      }
    } catch (n) {
      throw l.error("Error reading file stream:", n), n;
    }
  }
  /**
   * Legacy fallback for browsers without stream() support
   */
  static async readFileInChunksLegacy(e, t = {}) {
    const r = t.chunkSizeBytes || this.DEFAULT_CHUNK_SIZE, s = e.size;
    let a = 0, i = 0;
    for (; a < s && !t.signal?.aborted; ) {
      const n = Math.min(a + r, s), o = e.slice(a, n), u = await this.readBlobAsText(o, t.encoding), d = Math.min(100, n / s * 100), h = {
        chunkId: i++,
        text: u,
        isFirstChunk: a === 0,
        isLastChunk: n >= s,
        progress: d
      };
      t.onChunk && await t.onChunk(h), t.onProgress && t.onProgress(d), a = n;
    }
  }
  /**
   * Read blob as text using FileReader
   */
  static readBlobAsText(e, t) {
    return new Promise((r, s) => {
      const a = new FileReader();
      a.onload = () => {
        r(a.result);
      }, a.onerror = () => {
        s(new Error("Failed to read blob"));
      }, a.readAsText(e, t || "utf-8");
    });
  }
  /**
   * Get optimal chunk size based on file size
   */
  static getOptimalChunkSize(e) {
    return e < 1024 * 1024 ? 256 * 1024 : e < 10 * 1024 * 1024 ? 1024 * 1024 : e < 100 * 1024 * 1024 ? 2 * 1024 * 1024 : 5 * 1024 * 1024;
  }
};
X.DEFAULT_CHUNK_SIZE = 1024 * 1024;
let $ = X;
function he() {
  return typeof process < "u" && process.versions != null && process.versions.node != null;
}
function ze() {
  try {
    if (typeof __dirname < "u")
      return new Function("module", "return require(module)")("path").join(__dirname, "wasm", "csvParser.wasm");
  } catch (c) {
    l.warn("Could not resolve __dirname:", c);
  }
  return "./dist/wasm/csvParser.wasm";
}
function fe() {
  if (he())
    return ze();
  try {
    if (typeof import.meta < "u" && import.meta.url) {
      const c = import.meta.url, t = `${c.substring(0, c.lastIndexOf("/"))}/csvParser.wasm`;
      return l.info(`Computed WASM URL: ${t}`), t;
    }
  } catch (c) {
    l.warn("import.meta.url not available:", c);
  }
  return l.warn("Using fallback WASM path"), "/wasm/csvParser.wasm";
}
function Q(c) {
  return c.length >= 4 && c[0] === 0 && c[1] === 97 && c[2] === 115 && c[3] === 109;
}
function qe(c) {
  const e = new Function("module", "return require(module)"), t = e("fs"), r = e("path"), s = [
    c,
    r.resolve(c),
    r.join(__dirname, "wasm", "csvParser.wasm"),
    // For UMD build: dist/wasm/csvParser.wasm
    r.join(__dirname, "csvParser.wasm"),
    // For ESM build in wasm dir
    r.join(__dirname, "../wasm/csvParser.wasm"),
    // One level up
    r.join(
      process.cwd(),
      "node_modules/@mindfiredigital/pivothead/dist/wasm/csvParser.wasm"
    )
  ];
  for (const a of s)
    try {
      if (t.existsSync(a)) {
        l.info(`✅ Loading WASM from: ${a}`);
        const i = t.readFileSync(a), n = i.buffer.slice(
          i.byteOffset,
          i.byteOffset + i.byteLength
        ), o = new Uint8Array(n);
        if (Q(o))
          return n;
        l.warn(`Invalid WASM file at ${a}`);
      }
    } catch {
    }
  throw new Error(
    `WASM file not found. Tried paths:
${s.join(`
`)}

Make sure the package is properly installed.`
  );
}
async function Oe() {
  const c = fe();
  if (he())
    try {
      return qe(c);
    } catch (e) {
      throw l.error("Failed to load WASM in Node.js:", e), e;
    }
  try {
    const e = await fetch(c);
    if (!e.ok)
      throw new Error(`HTTP ${e.status}: ${e.statusText}`);
    const t = await e.arrayBuffer(), r = new Uint8Array(t);
    if (Q(r))
      return t;
    {
      const s = String.fromCharCode(...r.slice(0, 4));
      throw s === "<!DO" || s === "<htm" ? new Error(
        `Received HTML instead of WASM binary from ${c} - likely a 404 error`
      ) : new Error(`Invalid WASM file format from ${c}`);
    }
  } catch (e) {
    l.warn(`Failed to fetch WASM from ${c}:`, e);
    const t = [
      "/wasm/csvParser.wasm",
      "/node_modules/@mindfiredigital/pivothead/dist/wasm/csvParser.wasm"
    ];
    for (const r of t)
      if (r !== c)
        try {
          l.info(`Trying fallback URL: ${r}`);
          const s = await fetch(r);
          if (!s.ok) continue;
          const a = await s.arrayBuffer(), i = new Uint8Array(a);
          if (Q(i))
            return l.info(
              `✅ Successfully loaded WASM from fallback: ${r}`
            ), a;
        } catch (s) {
          l.warn(`Fallback ${r} failed:`, s);
        }
    throw new Error(
      `Failed to fetch WASM from ${c} and all fallback URLs. Make sure to run 'pnpm build:wasm' in the core package.`
    );
  }
}
let O = null;
async function Ve() {
  if (O) return O;
  try {
    return O = (await import("./index-DFEh6Cpe.mjs")).instantiate, O;
  } catch (c) {
    return l.warn("AssemblyScript loader not available:", c), null;
  }
}
const M = class M {
  constructor() {
    this.wasmModule = null, this.isLoaded = !1, this.loadPromise = null;
  }
  /**
   * Get singleton instance
   */
  static getInstance() {
    return M.instance || (M.instance = new M()), M.instance;
  }
  /**
   * Check if WebAssembly is supported
   */
  static isSupported() {
    try {
      return typeof WebAssembly < "u" && typeof WebAssembly.instantiate == "function";
    } catch {
      return !1;
    }
  }
  /**
   * Load the WASM module
   */
  async load() {
    if (!this.isLoaded)
      return this.loadPromise ? this.loadPromise : (this.loadPromise = this._loadModule(), this.loadPromise);
  }
  /**
   * Internal method to load WASM module
   */
  async _loadModule() {
    try {
      const e = await Ve();
      if (!e)
        throw new Error(
          "@assemblyscript/loader not available. Install with: pnpm add @assemblyscript/loader"
        );
      if (!M.isSupported())
        throw new Error("WebAssembly is not supported in this browser");
      l.info("🚀 Loading WebAssembly CSV parser...");
      const t = fe();
      l.info(`Loading WASM from: ${t}`);
      let r;
      try {
        r = await Oe(), l.info(
          `✅ WASM file loaded successfully (${r.byteLength} bytes)`
        );
      } catch (a) {
        throw l.error(
          `❌ Failed to fetch WASM file from ${t}:`,
          a
        ), new Error(
          `WASM file not found at ${t}. Make sure to run 'npm run build:wasm' in the core package.`
        );
      }
      this.wasmModule = await e(r, {
        env: {
          abort: (a, i, n, o) => {
            l.error(`WASM abort: ${a} at ${i}:${n}:${o}`);
          }
        }
      }), this.isLoaded = !0;
      const s = this.wasmModule.exports.getVersion();
      l.info(
        `✅ WebAssembly CSV parser loaded successfully (v${s})`
      );
    } catch (e) {
      throw this.isLoaded = !1, this.loadPromise = null, l.error("❌ Failed to load WASM module:", e), e;
    }
  }
  /**
   * Check if WASM module is loaded
   */
  isModuleLoaded() {
    return this.isLoaded;
  }
  /**
   * Parse CSV chunk using WASM
   */
  parseCSVChunk(e, t = {}) {
    if (!this.isLoaded || !this.wasmModule)
      throw new Error("WASM module not loaded. Call load() first.");
    const r = t.delimiter?.charCodeAt(0) ?? 44, s = t.hasHeader ?? !0, a = t.trimValues ?? !0;
    try {
      const i = this.wasmModule.exports.parseCSVChunk(
        this.wasmModule.exports.__newString(e),
        r,
        s ? 1 : 0,
        a ? 1 : 0
      ), n = this.wasmModule.exports.getLastRowCount(), o = this.wasmModule.exports.getLastColCount(), u = this.wasmModule.exports.getLastErrorMessage(), d = this.wasmModule.exports.__getString(u);
      return {
        rowCount: n,
        colCount: o,
        errorCode: i,
        errorMessage: d
      };
    } catch (i) {
      throw l.error("Error in WASM parseCSVChunk:", i), i;
    }
  }
  /**
   * Extract field from CSV data using WASM
   */
  extractField(e, t, r, s = !0) {
    if (!this.isLoaded || !this.wasmModule)
      throw new Error("WASM module not loaded. Call load() first.");
    const a = this.wasmModule.exports.extractField(
      this.wasmModule.exports.__newString(e),
      t,
      r,
      s ? 1 : 0
    );
    return this.wasmModule.exports.__getString(a);
  }
  /**
   * Parse number using WASM
   */
  parseNumber(e) {
    if (!this.isLoaded || !this.wasmModule)
      throw new Error("WASM module not loaded. Call load() first.");
    return this.wasmModule.exports.parseNumber(
      this.wasmModule.exports.__newString(e)
    );
  }
  /**
   * Detect field type using WASM
   * Returns: 0=string, 1=number, 2=boolean, 3=null
   */
  detectFieldType(e) {
    if (!this.isLoaded || !this.wasmModule)
      throw new Error("WASM module not loaded. Call load() first.");
    return this.wasmModule.exports.detectFieldType(
      this.wasmModule.exports.__newString(e)
    );
  }
  /**
   * Estimate memory usage for parsed data
   */
  estimateMemory(e, t) {
    if (!this.isLoaded || !this.wasmModule)
      throw new Error("WASM module not loaded. Call load() first.");
    return this.wasmModule.exports.estimateMemory(e, t);
  }
  /**
   * Get WASM module version
   */
  getVersion() {
    if (!this.isLoaded || !this.wasmModule)
      throw new Error("WASM module not loaded. Call load() first.");
    const e = this.wasmModule.exports.getVersion();
    return this.wasmModule.exports.__getString(e);
  }
  /**
   * Run benchmark
   */
  benchmark(e) {
    if (!this.isLoaded || !this.wasmModule)
      throw new Error("WASM module not loaded. Call load() first.");
    return this.wasmModule.exports.benchmark(
      this.wasmModule.exports.__newString(e)
    );
  }
  /**
   * Unload WASM module and free resources
   */
  unload() {
    this.wasmModule && this.wasmModule.exports.__release && this.wasmModule.exports.__release(), this.wasmModule = null, this.isLoaded = !1, this.loadPromise = null;
  }
};
M.instance = null;
let T = M;
const He = () => T.getInstance();
class Ue {
  constructor() {
    this.isInitialized = !1, this.wasmLoader = He();
  }
  /**
   * Initialize WASM module
   */
  async initialize() {
    if (this.isInitialized)
      return !0;
    try {
      return T.isSupported() ? (await this.wasmLoader.load(), this.isInitialized = !0, !0) : (l.warn(
        "⚠️ WebAssembly not supported, falling back to Web Workers"
      ), !1);
    } catch (e) {
      return l.warn("⚠️ Failed to initialize WASM:", e), !1;
    }
  }
  /**
   * Check if WASM is ready
   */
  isReady() {
    return this.isInitialized && this.wasmLoader.isModuleLoaded();
  }
  /**
   * Process CSV data using WASM
   */
  async processCSV(e, t = {}) {
    const r = performance.now();
    try {
      if (!this.isReady())
        throw new Error("WASM module not initialized");
      const {
        delimiter: s = ",",
        hasHeader: a = !0,
        trimValues: i = !0,
        skipEmptyLines: n = !0
      } = t;
      l.info("🚀 Processing CSV with WebAssembly...");
      const o = this.wasmLoader.parseCSVChunk(e, {
        delimiter: s,
        hasHeader: a,
        trimValues: i
      });
      if (o.errorCode !== 0)
        throw new Error(o.errorMessage || "WASM parsing failed");
      l.info(
        `📊 WASM parsed: ${o.rowCount} rows, ${o.colCount} columns`
      );
      const u = this.parseCSVRows(e, {
        delimiter: s,
        hasHeader: a,
        trimValues: i,
        skipEmptyLines: n
      });
      let d, h;
      t.providedHeaders && t.providedHeaders.length > 0 ? (d = t.providedHeaders, h = u) : a && u.length > 0 ? (d = u[0], h = u.slice(1)) : (d = void 0, h = u);
      const g = this.rowsToObjects(h, d), m = performance.now() - r;
      return l.info(`✅ WASM processing completed in ${m.toFixed(2)}ms`), {
        success: !0,
        data: g,
        headers: d,
        rowCount: h.length,
        colCount: o.colCount,
        parseTime: m
      };
    } catch (s) {
      const a = performance.now() - r;
      return l.error("❌ WASM processing error:", s), {
        success: !1,
        data: [],
        rowCount: 0,
        colCount: 0,
        parseTime: a,
        error: s instanceof Error ? s.message : "Unknown error"
      };
    }
  }
  /**
   * Parse CSV into rows (JavaScript implementation)
   * TODO: Move this logic to WASM for better performance
   */
  parseCSVRows(e, t) {
    const {
      delimiter: r = ",",
      trimValues: s = !0,
      skipEmptyLines: a = !0
    } = t, i = [];
    let n = [], o = "", u = !1;
    for (let d = 0; d < e.length; d++) {
      const h = e[d], g = d + 1 < e.length ? e[d + 1] : "";
      u ? h === '"' ? g === '"' ? (o += '"', d++) : u = !1 : o += h : h === '"' ? u = !0 : h === r ? (n.push(s ? o.trim() : o), o = "") : h === `
` || h === "\r" ? (h === "\r" && g === `
` && d++, n.push(s ? o.trim() : o), o = "", (!a || n.some((m) => m !== "")) && i.push(n), n = []) : o += h;
    }
    return (o !== "" || n.length > 0) && (n.push(s ? o.trim() : o), (!a || n.some((d) => d !== "")) && i.push(n)), i;
  }
  /**
   * Convert rows to objects
   */
  rowsToObjects(e, t) {
    return !t || t.length === 0 ? e.map((r) => ({ ...r })) : e.map((r) => {
      const s = {};
      return t.forEach((a, i) => {
        const n = r[i] ?? "";
        if (this.isReady() && n !== "")
          switch (this.wasmLoader.detectFieldType(n)) {
            case 1:
              s[a] = this.wasmLoader.parseNumber(n);
              break;
            case 2:
              s[a] = n.toLowerCase() === "true";
              break;
            case 3:
              s[a] = null;
              break;
            default:
              s[a] = n;
          }
        else
          s[a] = n;
      }), s;
    });
  }
  /**
   * Process CSV file using WASM
   */
  async processFile(e, t = {}) {
    try {
      if (e.size > q)
        return l.warn(
          `⚠️ File too large for WASM (${(e.size / 1024 / 1024).toFixed(2)}MB > 8MB). Should use Web Workers instead.`
        ), {
          success: !1,
          data: [],
          rowCount: 0,
          colCount: 0,
          parseTime: 0,
          error: `File too large for WASM processing (${(e.size / 1024 / 1024).toFixed(2)}MB). Falling back to Web Workers.`
        };
      const r = await e.text();
      return this.processCSV(r, t);
    } catch (r) {
      return {
        success: !1,
        data: [],
        rowCount: 0,
        colCount: 0,
        parseTime: 0,
        error: r instanceof Error ? r.message : "Unknown error"
      };
    }
  }
  /**
   * Estimate memory usage for a dataset
   */
  estimateMemory(e, t) {
    return this.isReady() ? this.wasmLoader.estimateMemory(e, t) : e * t * 64;
  }
  /**
   * Get WASM version
   */
  getVersion() {
    return this.isReady() ? this.wasmLoader.getVersion() : "N/A";
  }
  /**
   * Cleanup and release resources
   */
  cleanup() {
    this.wasmLoader && this.wasmLoader.unload(), this.isInitialized = !1;
  }
}
let U = null;
const ie = () => (U || (U = new Ue()), U);
function me(c) {
  if (c == null) return null;
  if (typeof c == "number") return isFinite(c) ? c : null;
  if (typeof c != "string") return null;
  let e = c.trim();
  if (!e) return null;
  const t = /[€£¥₹$]/.test(e), r = /[A-Za-z]/.test(e), s = /[.,]/.test(e), a = e.startsWith("(") && e.endsWith(")");
  if (r && !t || !t && !s && !a) return null;
  let i = !1;
  if (a && (i = !0, e = e.slice(1, -1)), e = e.replace(/[^0-9,.-]/g, ""), (e.match(/-/g) || []).length > 1) return null;
  e = e.replace(/(.*)-(.*)/, "-$1$2");
  const n = e.includes("."), o = e.includes(","), u = e.lastIndexOf("."), d = e.lastIndexOf(",");
  let h = e;
  if (n && o) {
    const m = u > d ? "." : ",", C = m === "." ? "," : ".";
    h = h.split(C).join(""), h = h.replace(m, ".");
  } else if (o && !n) {
    const m = e.split(",");
    if (m.length > 2)
      h = m.join("");
    else if (m.length === 2) {
      const C = m[1];
      C.length === 2 || C.length === 3 ? h = m[0] + "." + C : h = m.join("");
    }
  } else if (n && !o) {
    const m = e.split(".");
    if (m.length > 2) {
      const C = m.pop();
      h = m.join("") + "." + C;
    } else
      h = e;
  } else
    h = e;
  const g = Number(h);
  return isNaN(g) ? null : i ? -g : g;
}
function J(c, e) {
  const t = {}, r = c.slice(0, Math.min(200, c.length));
  for (const s of e) {
    let a = 0, i = 0, n = 0, o = 0, u = 0;
    for (const g of r) {
      const m = g?.[s];
      if (m == null || m === "") {
        u++;
        continue;
      }
      if (typeof m == "number") a++;
      else if (typeof m == "boolean") o++;
      else if (typeof m == "string") {
        if (me(m) !== null) {
          a++;
          continue;
        }
        const S = Date.parse(m);
        !isNaN(S) && /\d{4}-\d{2}-\d{2}|\//.test(m) ? n++ : isNaN(Number(m)) ? i++ : a++;
      } else m instanceof Date ? n++ : i++;
    }
    switch ((Object.entries({ num: a, str: i, date: n, bool: o, nul: u }).sort((g, m) => m[1] - g[1])[0] || [
      "str"
    ])[0]) {
      case "num":
        t[s] = "number";
        break;
      case "date":
        t[s] = "date";
        break;
      case "bool":
        t[s] = "boolean";
        break;
      case "str":
        t[s] = "string";
        break;
      default:
        t[s] = "unknown";
    }
  }
  return t;
}
function Ie(c, e) {
  const t = c.split(`
`).filter((n) => e.skipEmptyLines ? n.trim() !== "" : !0);
  if (t.length === 0) return [];
  const r = e.delimiter || ",";
  let s = [], a = 0;
  e.hasHeader ? (s = I(t[0], r), e.trimValues && (s = s.map((n) => n.trim())), a = 1) : s = I(t[0], r).map((o, u) => `Column_${u + 1}`);
  const i = [];
  for (let n = a; n < t.length; n++) {
    const o = I(t[n], r);
    if (o.length === 0) continue;
    const u = {};
    s.forEach((d, h) => {
      let g = o[h] || "";
      e.trimValues && (g = g.trim()), u[d] = Ge(g);
    }), i.push(u);
  }
  return i;
}
function I(c, e) {
  const t = [];
  let r = "", s = !1;
  for (let a = 0; a < c.length; a++) {
    const i = c[a];
    i === '"' && (a === 0 || c[a - 1] === e || s) ? s = !s : i === e && !s ? (t.push(r), r = "") : r += i;
  }
  return t.push(r), t;
}
function Ge(c) {
  if (c === "" || c === null || c === void 0)
    return null;
  const e = c.replace(/^"(.*)"$/, "$1"), t = me(e);
  if (t !== null) return t;
  if (!isNaN(Number(e)) && e !== "")
    return Number(e);
  if (e.toLowerCase() === "true") return !0;
  if (e.toLowerCase() === "false") return !1;
  const r = Date.parse(e);
  return !isNaN(r) && (e.match(/^\d{4}-\d{2}-\d{2}/) || e.includes("/")) ? new Date(r).toISOString().split("T")[0] : e;
}
function je(c) {
  if (!c || c.length === 0) return c;
  const e = JSON.parse(JSON.stringify(c)), t = Object.keys(e[0]), r = J(e, t);
  for (const s of e)
    for (const a of t)
      if (r[a] === "number" && typeof s[a] == "string") {
        const i = parseFloat(s[a]);
        isNaN(i) || (s[a] = i);
      }
  return e;
}
const V = [".csv", ".txt"], H = [".json", ".jsonl"], Be = 1024 * 1024 * 1024;
function G(c) {
  return new Promise((e) => {
    const t = document.createElement("input");
    t.type = "file", t.accept = c.join(","), t.style.display = "none", t.onchange = (r) => {
      const a = r.target.files?.[0] || null;
      document.body.removeChild(t), e(a);
    }, t.oncancel = () => {
      document.body.removeChild(t), e(null);
    }, document.body.appendChild(t), t.click();
  });
}
function ne(c, e) {
  const t = e.maxFileSize || Be;
  if (c.size > t)
    return {
      success: !1,
      error: `File size (${A(c.size)}) exceeds maximum allowed size (${A(t)})`
    };
  const r = ge(c.name), s = [
    ...V,
    ...H
  ];
  return s.includes(r) ? { success: !0 } : {
    success: !1,
    error: `Unsupported file type: ${r}. Supported types: ${s.join(", ")}`
  };
}
function oe(c, e) {
  return new Promise((t, r) => {
    const s = new FileReader();
    s.onload = () => {
      e && e(100), t(s.result);
    }, s.onerror = () => r(new Error("Failed to read file")), e && (s.onprogress = (a) => {
      if (a.lengthComputable) {
        const i = a.loaded / a.total * 100;
        e(i);
      }
    }), s.readAsText(c);
  });
}
function ge(c) {
  return c.toLowerCase().substring(c.lastIndexOf("."));
}
function A(c) {
  if (c === 0) return "0 Bytes";
  const e = 1024, t = ["Bytes", "KB", "MB", "GB"], r = Math.floor(Math.log(c) / Math.log(e));
  return parseFloat((c / Math.pow(e, r)).toFixed(2)) + " " + t[r];
}
function Ke(c, e) {
  const t = [];
  if (e.length === 0)
    return t.push("No columns detected in data"), t;
  const r = e.filter((n) => !n || n.trim() === "");
  r.length > 0 && t.push(`Found ${r.length} columns with empty names`);
  const s = e.filter(
    (n, o) => e.indexOf(n) !== o
  );
  s.length > 0 && t.push(
    `Found duplicate column names: ${[...new Set(s)].join(", ")}`
  );
  const a = Math.min(100, c.length), i = [];
  for (let n = 0; n < a; n++) {
    const o = c[n];
    Object.keys(o).length !== e.length && i.push(n + 1);
  }
  return i.length > 0 && t.push(
    `Found ${i.length} rows with inconsistent column count in sample`
  ), t;
}
function Qe(c, e) {
  const t = e.split(".");
  let r = c;
  for (const s of t)
    if (r && typeof r == "object" && r !== null && s in r)
      r = r[s];
    else
      return [];
  return Array.isArray(r) ? r : [];
}
const ee = class ee {
  /**
   * Get current performance configuration
   */
  static getConfig() {
    return { ...this.config };
  }
  /**
   * Update performance configuration
   */
  static updateConfig(e) {
    this.config = { ...this.config, ...e };
  }
  /**
   * Check if workers should be used for given size/rows
   */
  static shouldUseWorkers(e, t) {
    return t !== void 0 ? e >= this.config.useWorkersAboveSize || t >= this.config.useWorkersAboveRows : e >= this.config.useWorkersAboveSize;
  }
  /**
   * Check if WASM should be used for given size/rows
   */
  static shouldUseWasm(e, t) {
    return t !== void 0 ? e >= this.config.useWasmAboveSize || t >= this.config.useWasmAboveRows : e >= this.config.useWasmAboveSize;
  }
  /**
   * Check if drag/drop should be enabled
   */
  static isDragDropAllowed(e) {
    return e <= this.config.maxRowsForDragDrop;
  }
  /**
   * Check if pagination is required
   */
  static requiresPagination(e) {
    return e > this.config.maxRowsForFullRender;
  }
  /**
   * Get appropriate page size for dataset
   */
  static getPageSize(e) {
    return e <= 100 || e <= 1e3 ? 50 : (e <= 1e4, 100);
  }
  /**
   * Check if sampling should be used
   */
  static shouldSample(e) {
    return e > this.config.sampleThreshold;
  }
  /**
   * Get sample size for dataset
   */
  static getSampleSize(e) {
    return this.shouldSample(e) ? Math.min(this.config.sampleSize, e) : e;
  }
  /**
   * Get optimal chunk size for file
   */
  static getChunkSize(e) {
    return e < ae ? Ce : e < ke ? ae : e < Fe ? De : ve;
  }
  /**
   * Get performance mode for dataset
   * Returns the optimal processing mode based on file size
   */
  static getPerformanceMode(e, t) {
    return e > q ? "streaming-wasm" : e >= j && e <= q ? "wasm" : this.shouldUseWorkers(e, t) ? "workers" : "standard";
  }
  /**
   * Check if virtual scrolling should be used
   * @param rowCount Number of rows in the dataset
   * @returns true if virtual scrolling should be enabled
   */
  static shouldUseVirtualScroll(e) {
    return this.config.enableVirtualScroll && e >= this.config.virtualScrollThreshold;
  }
  /**
   * Get virtual scroll configuration
   * @returns Virtual scroll settings
   */
  static getVirtualScrollConfig() {
    return {
      rowHeight: this.config.virtualScrollRowHeight,
      bufferSize: this.config.virtualScrollBuffer,
      threshold: this.config.virtualScrollThreshold
    };
  }
};
ee.config = {
  // Worker configuration - for files < 5MB
  useWorkersAboveSize: ue,
  useWorkersAboveRows: 5e3,
  // 5k rows
  // WASM configuration - for files >= 5MB (up to 8MB safety limit)
  // Strategy: Workers for < 5MB, WASM for 5-8MB, Workers for > 8MB (streaming)
  useWasmAboveSize: j,
  useWasmAboveRows: 2e4,
  // 20k rows
  // UI interaction limits (with virtual scrolling support)
  maxRowsForDragDrop: 1e3,
  // Traditional rendering limit (virtual scrolling handles larger datasets)
  maxRowsForFullRender: 1e4,
  // Virtual scrolling threshold
  defaultPageSize: 50,
  // Show 50 rows per page
  maxPageSize: 500,
  // Max 500 rows per page
  // Sampling configuration
  sampleThreshold: 1e4,
  // Sample if > 10k rows
  sampleSize: 5e3,
  // Sample 5k rows
  // Virtual scrolling configuration
  enableVirtualScroll: !0,
  // Enable virtual scrolling
  virtualScrollThreshold: 1e3,
  // Auto-enable at 1k rows
  virtualScrollRowHeight: 40,
  // Default row height (px)
  virtualScrollBuffer: 10
  // Extra rows for smooth scrolling
};
let z = ee;
function Je(c) {
  const e = c.length > 0 ? Object.keys(c[0]) : [];
  let t = c;
  if (e.length === 1) {
    const [f] = e, p = [];
    if (J(t, e)[f] === "number") {
      for (let F = 0; F < t.length; F++)
        t[F].__all__ = "All";
      const k = [{ uniqueName: "__all__", caption: "All" }], v = [
        { uniqueName: "__all__", caption: "All" }
      ];
      return p.push({
        uniqueName: f,
        caption: `Sum of ${f}`,
        aggregation: "sum",
        format: {
          type: "number",
          decimals: 2,
          locale: "en-US"
        }
      }), {
        rows: k,
        columns: v,
        measures: p,
        data: t,
        columnsList: ["__all__", ...e]
      };
    }
    for (let k = 0; k < t.length; k++)
      t[k].__all__ = "All";
    return {
      rows: [{ uniqueName: f, caption: f }],
      columns: [
        { uniqueName: "__all__", caption: "All" }
      ],
      measures: p,
      data: t,
      columnsList: ["__all__", ...e]
    };
  }
  const r = J(t, e), s = e.filter((f) => r[f] === "number"), a = e.filter((f) => r[f] !== "number"), i = s.map((f) => {
    const p = t.some((w) => {
      const y = w[f];
      return typeof y == "string" && /[$€£¥₹]/.test(y);
    });
    return {
      uniqueName: f,
      caption: `Sum of ${f}`,
      aggregation: "sum",
      format: {
        type: p ? "currency" : "number",
        currency: "USD",
        locale: "en-US",
        decimals: 2
      },
      sortabled: !0
    };
  });
  let n = [], o = [];
  const u = t.length || 1, d = u > 1e4, h = d ? Math.min(5e3, u) : u, g = d ? t.slice(0, h) : t, m = (f) => {
    const p = new Set(g.map((y) => y?.[f])).size;
    if (!d) return p;
    const w = Math.round(p * (u / h));
    return Math.min(w, u);
  }, C = (f) => new Set(g.map((p) => p?.[f])).size;
  if (a.length === 0) {
    for (let f = 0; f < t.length; f++)
      t[f].__all__ = "All";
    return n = [{ uniqueName: "__all__", caption: "All" }], o = [{ uniqueName: "__all__", caption: "All" }], {
      rows: n,
      columns: o,
      measures: i,
      data: t,
      columnsList: ["__all__", ...e]
    };
  }
  const S = a.map((f) => ({ f, u: m(f) })).sort((f, p) => f.u - p.u);
  l.info("DEBUG: Non-numeric fields:", a), l.info("DEBUG: Ranked fields:", S), l.info("DEBUG: Total rows:", u);
  const b = S.filter((f) => {
    const p = C(f.f);
    return f.u > 1 && f.u < u && p <= Ne;
  });
  if (l.info(
    "DEBUG: Candidate dimensions (filtered by max cardinality):",
    b
  ), l.info(
    "DEBUG: Detailed ranked fields:",
    S.map((f) => ({ field: f.f, unique: f.u }))
  ), b.length >= 2) {
    const f = b.sort((p, w) => {
      if (p.u !== w.u) return p.u - w.u;
      const y = p.f.toLowerCase().trim(), D = w.f.toLowerCase().trim(), k = Ee, v = Ae, F = k.some((N) => y.includes(N)), _ = k.some((N) => D.includes(N)), L = v.some((N) => y.includes(N)), R = v.some((N) => D.includes(N));
      return l.info("DEBUG: Field comparison:", {
        aField: y,
        bField: D,
        aIsRow: F,
        bIsRow: _,
        aIsColumn: L,
        bIsColumn: R
      }), F && !_ ? -1 : _ && !F || L && !R ? 1 : R && !L ? -1 : y.localeCompare(D);
    });
    l.info(
      "DEBUG: Sorted candidates:",
      f.map((p) => ({ field: p.f, unique: p.u }))
    ), n = [
      { uniqueName: f[0].f, caption: f[0].f }
    ], o = [
      { uniqueName: f[1].f, caption: f[1].f }
    ], l.info(
      "DEBUG: Final assignment - Rows:",
      n,
      "Columns:",
      o
    );
  } else if (b.length === 1)
    n = [
      { uniqueName: b[0].f, caption: b[0].f.trim() }
    ];
  else if (a.length >= 2) {
    const f = S.sort((p, w) => {
      if (p.u !== w.u) return p.u - w.u;
      const y = p.f.toLowerCase().trim(), D = w.f.toLowerCase().trim(), k = [
        "region",
        "country",
        "state",
        "city",
        "category",
        "department"
      ], v = [
        "product",
        "item",
        "month",
        "quarter",
        "year",
        "type"
      ], F = k.some((N) => y.includes(N)), _ = k.some((N) => D.includes(N)), L = v.some((N) => y.includes(N)), R = v.some((N) => D.includes(N));
      return F && !_ ? -1 : _ && !F || L && !R ? 1 : R && !L ? -1 : y.localeCompare(D);
    });
    n = [
      {
        uniqueName: f[0].f,
        caption: f[0].f.trim()
      }
    ], o = [
      {
        uniqueName: f[1].f,
        caption: f[1].f.trim()
      }
    ];
  } else
    n = [{ uniqueName: S[0].f, caption: S[0].f.trim() }];
  if (!o || o.length === 0) {
    for (let f = 0; f < t.length; f++)
      t[f].__all__ = "All";
    o = [{ uniqueName: "__all__", caption: "All" }];
  }
  return {
    rows: n,
    columns: o,
    measures: i,
    data: t,
    columnsList: Object.keys(t[0] || {})
  };
}
const Z = 5e4, Y = 2e4;
function W(c) {
  const { file: e, engine: t, options: r, performanceModeOverride: s, parseTime: a } = c;
  let i = c.parsedData;
  const n = e.size > B, o = r.maxRecords ?? (n ? Y : Z), u = i.length;
  if (i.length > o && (l.warn(
    `⚠️ Dataset has ${u.toLocaleString()} rows. Loading first ${o.toLocaleString()} rows for performance.`
  ), i = i.slice(0, o)), i.length === 0)
    return {
      success: !1,
      error: "No valid data found in file"
    };
  const d = Object.keys(i[0]), h = Ke(i, d);
  h.length > 0 && l.warn("Data validation warnings:", h), t.setAutoAllColumn(!0);
  const g = Je(i), { rows: m, columns: C, measures: S, data: b } = g;
  t.updateDataSource(b), t.setLayout(m, C, S);
  try {
    "setConditionalFormatting" in t && typeof t.setConditionalFormatting == "function" && t.setConditionalFormatting([
      {
        value: {
          type: "Number",
          operator: "Greater than",
          value1: 1e3
        },
        format: {
          backgroundColor: "#d4edda",
          color: "#155724"
        }
      },
      {
        value: {
          type: "Number",
          operator: "Less than",
          value1: 0
        },
        format: {
          backgroundColor: "#f8d7da",
          color: "#721c24"
        }
      }
    ]);
  } catch (y) {
    l.warn(
      "Conditional formatting not supported by this engine version:",
      y
    );
  }
  u > o && h.push(
    `Note: Displaying ${o.toLocaleString()} of ${u.toLocaleString()} rows for optimal performance.`
  );
  const f = s ?? z.getPerformanceMode(e.size, b.length), p = z.isDragDropAllowed(
    b.length
  ), w = z.requiresPagination(
    b.length
  );
  return p || l.info(
    `Large dataset detected: ${b.length.toLocaleString()} rows. Virtual scrolling will be used for optimal performance.`
  ), {
    success: !0,
    data: b,
    fileName: e.name,
    fileSize: e.size,
    recordCount: b.length,
    columns: d,
    validationErrors: h.length > 0 ? h : void 0,
    performanceMode: f,
    allowDragDrop: p,
    requiresPagination: w,
    parseTime: a
  };
}
function Ze() {
  return `
/* CSV Parser Worker */
function parseCsvLine(line, delimiter) {
  const result = [];
  let current = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    if (char === '"' && (i === 0 || line[i - 1] === delimiter || inQuotes)) {
      inQuotes = !inQuotes;
    } else if (char === delimiter && !inQuotes) {
      result.push(current);
      current = '';
    } else {
      current += char;
    }
  }
  result.push(current);
  return result;
}

function convertValue(value) {
  if (value === '' || value === null || value === undefined) return null;
  const cleaned = value.replace(/^"(.*)"$/, '$1').trim();
  if (!isNaN(Number(cleaned)) && cleaned !== '') return Number(cleaned);
  if (cleaned.toLowerCase() === 'true') return true;
  if (cleaned.toLowerCase() === 'false') return false;
  const dateValue = Date.parse(cleaned);
  if (!isNaN(dateValue) && (cleaned.match(/^\\d{4}-\\d{2}-\\d{2}/) || cleaned.includes('/'))) {
    return new Date(dateValue).toISOString().split('T')[0];
  }
  return cleaned;
}

function parseChunk(text, delimiter, headers, leftover, isFirstChunk, isLastChunk) {
  const fullText = (leftover || '') + text;
  const lines = fullText.split('\\n');
  const incompleteLine = !isLastChunk ? lines.pop() : '';
  const data = [];
  let processedHeaders = headers;
  let startIndex = 0;

  if (isFirstChunk && !processedHeaders) {
    if (lines.length > 0) {
      processedHeaders = parseCsvLine(lines[0], delimiter).map(h => h.trim());
      startIndex = 1;
    }
  }

  for (let i = startIndex; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;
    const values = parseCsvLine(line, delimiter);
    if (values.length === 0) continue;
    const row = {};
    processedHeaders?.forEach((header, index) => {
      row[header] = convertValue(values[index] || '');
    });
    data.push(row);
  }

  return { data, headers: isFirstChunk ? processedHeaders : undefined, leftover: incompleteLine };
}

self.onmessage = (event) => {
  const { type, chunkId, text, isFirstChunk, isLastChunk, delimiter, headers, leftover } = event.data;
  if (type === 'PARSE_CHUNK') {
    try {
      self.postMessage({ type: 'PROGRESS', chunkId, progress: 0 });
      const result = parseChunk(text, delimiter, headers, leftover, isFirstChunk, isLastChunk);
      self.postMessage({
        type: 'CHUNK_PARSED',
        chunkId,
        data: result.data,
        headers: result.headers,
        leftover: result.leftover,
        rowCount: result.data.length,
      });
    } catch (error) {
      self.postMessage({
        type: 'CHUNK_PARSED',
        chunkId,
        data: [],
        rowCount: 0,
        error: error.message || 'Unknown error',
      });
    }
  }
};
`;
}
const te = class te {
  /**
   * Opens a file picker for CSV files and imports data into the pivot engine
   */
  static async connectToLocalCSV(e, t = {}) {
    try {
      const r = await G(V);
      return r ? await this.processCSVFile(r, e, t) : { success: !1, error: "No file selected" };
    } catch (r) {
      return l.error("Error connecting to CSV file:", r), {
        success: !1,
        error: r instanceof Error ? r.message : "Unknown error occurred"
      };
    }
  }
  /**
   * Opens a file picker for JSON files and imports data into the pivot engine
   */
  static async connectToLocalJSON(e, t = {}) {
    try {
      const r = await G(H);
      return r ? await this.processJSONFile(r, e, t) : { success: !1, error: "No file selected" };
    } catch (r) {
      return l.error("Error connecting to JSON file:", r), {
        success: !1,
        error: r instanceof Error ? r.message : "Unknown error occurred"
      };
    }
  }
  /**
   * Generic method to connect to any supported file type
   */
  static async connectToLocalFile(e, t = {}) {
    try {
      const r = [
        ...V,
        ...H
      ], s = await G(r);
      if (!s)
        return { success: !1, error: "No file selected" };
      const a = ge(s.name);
      return V.includes(a) ? await this.processCSVFile(s, e, t) : H.includes(a) ? await this.processJSONFile(s, e, t) : {
        success: !1,
        error: `Unsupported file type: ${a}`
      };
    } catch (r) {
      return l.error("Error connecting to local file:", r), {
        success: !1,
        error: r instanceof Error ? r.message : "Unknown error occurred"
      };
    }
  }
  /**
   * Processes a CSV file and updates the pivot engine
   * Routes to the appropriate tier based on file size.
   */
  static async processCSVFile(e, t, r) {
    const s = ne(e, r);
    if (!s.success)
      return s;
    if (e.size > q && T.isSupported()) {
      l.info(
        `🚀 Large file detected (${A(e.size)}). Using Streaming + WASM hybrid mode...`
      );
      try {
        const a = await this.processCSVFileWithStreamingWasm(
          e,
          t,
          r
        );
        if (a.success) return a;
        l.warn("⚠️ Streaming + WASM failed, falling back to Workers");
      } catch (a) {
        l.warn("⚠️ Streaming + WASM error:", a), l.warn("Falling back to Web Workers");
      }
    }
    if (e.size >= j && e.size <= q && T.isSupported()) {
      l.info(
        `🚀 Medium-large file detected (${A(e.size)}). Using WASM for maximum speed...`
      );
      try {
        const a = await this.processCSVFileWithWasm(
          e,
          t,
          r
        );
        if (a.success) return a;
        l.warn("⚠️ WASM processing failed, falling back to Workers");
      } catch (a) {
        l.warn("⚠️ WASM processing error:", a), l.warn("Falling back to Web Workers");
      }
    }
    if (e.size >= ue && r.useWorkers !== !1)
      return l.info(
        `Medium file detected (${A(e.size)}). Using Web Workers for parallel processing.`
      ), this.processCSVFileWithWorkers(e, t, r);
    try {
      const a = {
        delimiter: ",",
        hasHeader: !0,
        skipEmptyLines: !0,
        trimValues: !0,
        ...r.csv
      }, i = await oe(e, r.onProgress), n = Ie(i, a), o = je(n);
      return W({
        file: e,
        parsedData: o,
        engine: t,
        options: r
      });
    } catch (a) {
      return {
        success: !1,
        error: `Failed to parse CSV file: ${a instanceof Error ? a.message : "Unknown error"}`
      };
    }
  }
  /**
   * Processes a CSV file using WebAssembly for maximum performance
   */
  static async processCSVFileWithWasm(e, t, r) {
    try {
      const s = {
        delimiter: ",",
        hasHeader: !0,
        skipEmptyLines: !0,
        trimValues: !0,
        ...r.csv
      }, a = ie();
      if (!await a.initialize())
        return l.warn(
          "⚠️ WASM initialization failed, falling back to Web Workers"
        ), { success: !1, error: "WASM initialization failed" };
      const n = await a.processFile(e, {
        delimiter: s.delimiter,
        hasHeader: s.hasHeader,
        trimValues: s.trimValues
      });
      return n.success ? W({
        file: e,
        parsedData: n.data,
        engine: t,
        options: r,
        performanceModeOverride: "wasm"
      }) : { success: !1, error: "WASM processing failed" };
    } catch (s) {
      return l.error("WASM processing error:", s), {
        success: !1,
        error: `WASM processing failed: ${s instanceof Error ? s.message : "Unknown error"}`
      };
    }
  }
  /**
   * Processes large CSV files using Streaming + WASM chunks
   * Optimized for files > 8MB
   */
  static async processCSVFileWithStreamingWasm(e, t, r) {
    try {
      l.info("🚀 Using Streaming + WASM hybrid mode for large file...");
      const s = {
        delimiter: ",",
        hasHeader: !0,
        skipEmptyLines: !0,
        trimValues: !0,
        ...r.csv
      }, a = ie();
      if (!await a.initialize())
        return l.warn(
          "⚠️ WASM initialization failed, falling back to Web Workers"
        ), { success: !1, error: "WASM initialization failed" };
      const n = e.size > B, o = r.maxRecords ?? (n ? Y : Z), u = [];
      let d, h = "", g = 0, m = 0;
      const C = $.getOptimalChunkSize(e.size);
      l.info(
        `Processing with WASM, chunk size: ${A(C)}, max records: ${o.toLocaleString()}`
      );
      const S = performance.now(), b = new AbortController();
      await $.readFileInChunks(e, {
        chunkSizeBytes: C,
        encoding: s.encoding,
        signal: b.signal,
        onProgress: (p) => {
          r.onProgress && r.onProgress(p);
        },
        onChunk: async (p) => {
          if (u.length >= o) {
            b.abort();
            return;
          }
          g++;
          const w = h + p.text;
          h = "";
          const y = await a.processCSV(w, {
            delimiter: s.delimiter,
            hasHeader: p.isFirstChunk && s.hasHeader,
            trimValues: s.trimValues,
            // For non-first chunks pass the already-detected headers so
            // rowsToObjects can map column values correctly (not numeric indices)
            providedHeaders: !p.isFirstChunk && d ? d : void 0
          });
          if (!y.success || !y.data) {
            l.warn(
              `⚠️ WASM failed on chunk ${p.chunkId}, chunk skipped`
            );
            return;
          }
          if (p.isFirstChunk && y.data.length > 0 && (d = Object.keys(y.data[0]), l.info(`📋 Headers detected: ${d.join(", ")}`)), !p.isLastChunk) {
            const v = w.lastIndexOf(`
`);
            v !== -1 && (h = w.substring(v + 1), y.data.length > 0 && y.data.pop());
          }
          m += y.data.length;
          const D = o - u.length, k = y.data.length <= D ? y.data : y.data.slice(0, D);
          for (let v = 0; v < k.length; v++)
            u.push(k[v]);
          l.info(
            `Chunk ${p.chunkId} processed: ${y.data.length} rows (collected: ${u.length}/${o})`
          ), y.data = [], u.length >= o && (l.info(
            `✅ Reached record limit (${o.toLocaleString()}), stopping file read early.`
          ), b.abort());
        }
      });
      const f = performance.now() - S;
      return l.info(
        `✅ Streaming + WASM completed in ${(f / 1e3).toFixed(2)}s`
      ), l.info(
        `Processed ${g} chunks, ${m} rows parsed, ${u.length} rows collected.`
      ), r.onProgress && r.onProgress(90), W({
        file: e,
        parsedData: u,
        engine: t,
        options: r,
        performanceModeOverride: "streaming-wasm",
        parseTime: f
      });
    } catch (s) {
      return l.error("Error in Streaming + WASM processing:", s), {
        success: !1,
        error: `Streaming + WASM failed: ${s instanceof Error ? s.message : "Unknown error"}`
      };
    }
  }
  /**
   * Processes a CSV file using Web Workers and streaming
   */
  static async processCSVFileWithWorkers(e, t, r) {
    try {
      const s = {
        delimiter: ",",
        hasHeader: !0,
        skipEmptyLines: !0,
        trimValues: !0,
        ...r.csv
      };
      if (!this.workerPool) {
        const S = Ze(), b = new Blob([S], { type: "application/javascript" }), f = URL.createObjectURL(b);
        this.workerPool = new We(
          f,
          r.workerCount || Math.max(1, navigator.hardwareConcurrency - 1)
        );
      }
      const a = e.size > B, i = r.maxRecords ?? (a ? Y : Z), n = [];
      let o, u, d = 0;
      const h = r.chunkSizeBytes || $.getOptimalChunkSize(e.size);
      l.info(
        `Processing with ${this.workerPool.getWorkerCount()} workers, chunk size: ${A(h)}, max records: ${i.toLocaleString()}`
      );
      const g = performance.now(), m = new AbortController();
      await $.readFileInChunks(e, {
        chunkSizeBytes: h,
        encoding: s.encoding,
        signal: m.signal,
        onProgress: (S) => {
          r.onProgress && r.onProgress(S);
        },
        onChunk: async (S) => {
          if (n.length >= i) {
            m.abort();
            return;
          }
          d++;
          const b = await this.workerPool.execute({
            type: "PARSE_CHUNK",
            chunkId: S.chunkId,
            text: S.text,
            isFirstChunk: S.isFirstChunk,
            isLastChunk: S.isLastChunk,
            delimiter: s.delimiter,
            headers: o,
            leftover: u
          });
          if (b.headers && (o = b.headers), u = b.leftover, b.data && b.data.length > 0) {
            const f = i - n.length, p = Math.min(b.data.length, f);
            for (let w = 0; w < p; w++)
              n.push(b.data[w]);
          }
          l.info(
            `Chunk ${S.chunkId} processed: ${b.rowCount} rows (collected: ${n.length}/${i})`
          ), n.length >= i && (l.info("✅ Reached record limit, stopping file read early."), m.abort());
        }
      });
      const C = performance.now() - g;
      return l.info(`Parsing completed in ${(C / 1e3).toFixed(2)}s`), l.info(
        `Streaming complete. Processed ${d} chunks, ${n.length} rows collected.`
      ), r.onProgress && r.onProgress(90), W({
        file: e,
        parsedData: n,
        engine: t,
        options: r,
        parseTime: C
      });
    } catch (s) {
      return l.error("Error in worker-based CSV processing:", s), {
        success: !1,
        error: `Failed to parse CSV file: ${s instanceof Error ? s.message : "Unknown error"}`
      };
    }
  }
  /**
   * Processes a JSON file and updates the pivot engine
   */
  static async processJSONFile(e, t, r) {
    const s = ne(e, r);
    if (!s.success)
      return s;
    try {
      const a = await oe(e, r.onProgress), i = JSON.parse(a);
      let n = [];
      if (r.json?.arrayPath)
        n = Qe(i, r.json.arrayPath);
      else if (Array.isArray(i))
        n = i;
      else if (i.data && Array.isArray(i.data))
        n = i.data;
      else if (i.records && Array.isArray(i.records))
        n = i.records;
      else
        for (const o in i)
          if (Array.isArray(i[o])) {
            n = i[o];
            break;
          }
      return n.length === 0 ? {
        success: !1,
        error: "No array data found in JSON file. Please ensure the file contains an array of objects."
      } : W({
        file: e,
        parsedData: n,
        engine: t,
        options: r
      });
    } catch (a) {
      return a instanceof SyntaxError ? {
        success: !1,
        error: "Invalid JSON format. Please check your file syntax."
      } : {
        success: !1,
        error: `Failed to parse JSON file: ${a instanceof Error ? a.message : "Unknown error"}`
      };
    }
  }
  /**
   * Creates a summary of the import result
   */
  static createImportSummary(e) {
    if (!e.success)
      return `Import failed: ${e.error}`;
    let t = `Successfully imported ${e.recordCount} records from ${e.fileName}`;
    return e.columns && (t += `
Columns: ${e.columns.join(", ")}`), e.fileSize && (t += `
File size: ${A(e.fileSize)}`), e.validationErrors && e.validationErrors.length > 0 && (t += `

Warnings:
${e.validationErrors.map((r) => `• ${r}`).join(`
`)}`), t;
  }
  /**
   * Shows a browser notification for import results
   */
  static showImportNotification(e) {
    const t = this.createImportSummary(e);
    if ("Notification" in window) {
      if (Notification.permission === "granted") {
        new Notification("PivotHead Import", { body: t });
        return;
      }
      if (Notification.permission !== "denied") {
        Notification.requestPermission().then((r) => {
          r === "granted" ? new Notification("PivotHead Import", { body: t }) : l.info(t);
        });
        return;
      }
    }
    l.info(t);
  }
};
te.workerPool = null;
let le = te;
class rt {
  constructor(e) {
    this.scrollTop = 0, this.visibleStart = 0, this.visibleEnd = 0, this.containerHeight = 0, this.wrapper = null, this.spacer = null, this.content = null, this.resizeObserver = null, this.scrollRAF = null, this.container = e.container, this.data = e.data || [], this.rowHeight = e.rowHeight || Me, this.bufferSize = e.bufferSize || _e, this.renderRow = e.renderRow, this.renderHeader = e.renderHeader, this.onVisibleRangeChange = e.onVisibleRangeChange, this.onDragDrop = e.onDragDrop, this.init();
  }
  /**
   * Initialize the virtual scrolling structure
   * @private
   */
  init() {
    if (!this.container) {
      l.error("VirtualScrollManager: Container not found");
      return;
    }
    this.wrapper = document.createElement("div"), this.wrapper.className = "virtual-scroller-wrapper", this.wrapper.style.cssText = `
      position: relative;
      overflow-y: auto;
      overflow-x: auto;
      height: 600px;
      border: 1px solid #dee2e6;
    `, this.spacer = document.createElement("div"), this.spacer.className = "virtual-scroller-spacer", this.spacer.style.cssText = `
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      pointer-events: none;
    `, this.content = document.createElement("div"), this.content.className = "virtual-scroller-content", this.content.style.cssText = `
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
    `, this.wrapper.appendChild(this.spacer), this.wrapper.appendChild(this.content), this.container.appendChild(this.wrapper), this.wrapper.addEventListener("scroll", () => this.handleScroll()), this.resizeObserver = new ResizeObserver(() => {
      this.containerHeight = this.wrapper.clientHeight, this.render();
    }), this.resizeObserver.observe(this.wrapper), this.containerHeight = this.wrapper.clientHeight;
  }
  /**
   * Handle scroll events with requestAnimationFrame throttling
   * @private
   */
  handleScroll() {
    this.scrollRAF === null && (this.scrollRAF = requestAnimationFrame(() => {
      this.scrollRAF = null, this.onScroll();
    }));
  }
  /**
   * Process scroll event
   * @private
   */
  onScroll() {
    if (!this.wrapper) return;
    const e = this.wrapper.scrollTop;
    Math.abs(e - this.scrollTop) > this.rowHeight / 2 && (this.scrollTop = e, this.render());
  }
  /**
   * Calculate the range of visible rows
   * @returns Object with start and end indices
   */
  getVisibleRange() {
    const e = Math.floor(this.scrollTop / this.rowHeight), t = Math.ceil(
      (this.scrollTop + this.containerHeight) / this.rowHeight
    ), r = Math.max(0, e - this.bufferSize), s = Math.min(this.data.length, t + this.bufferSize);
    return { start: r, end: s };
  }
  /**
   * Render the visible rows
   */
  render() {
    if (!this.data || this.data.length === 0 || !this.content || !this.spacer) {
      this.content && (this.content.innerHTML = '<div style="padding: 20px;">No data available</div>');
      return;
    }
    const { start: e, end: t } = this.getVisibleRange();
    this.visibleStart = e, this.visibleEnd = t, this.onVisibleRangeChange && this.onVisibleRangeChange(e, t);
    const r = this.data.length * this.rowHeight + Le;
    this.spacer.style.height = `${r}px`, this.content.innerHTML = "";
    const s = document.createElement("table");
    if (s.style.cssText = `
      width: 100%;
      border-collapse: collapse;
      table-layout: fixed;
    `, this.renderHeader) {
      const i = this.renderHeader();
      i.style.cssText = `
        position: sticky;
        top: 0;
        z-index: ${Re};
        background: ${xe};
      `, s.appendChild(i);
    }
    const a = document.createElement("tbody");
    a.style.transform = `translateY(${e * this.rowHeight}px)`;
    for (let i = e; i < t && !(i >= this.data.length); i++) {
      const n = this.renderRow(this.data[i], i);
      this.onDragDrop && this.addDragDropSupport(n, i), a.appendChild(n);
    }
    s.appendChild(a), this.content.appendChild(s);
  }
  /**
   * Add drag-and-drop support to a row
   * @private
   */
  addDragDropSupport(e, t) {
    e.setAttribute("draggable", "true"), e.style.cursor = "move", e.dataset.virtualIndex = String(t), e.addEventListener("dragstart", (r) => {
      r.dataTransfer && (r.dataTransfer.effectAllowed = "move", r.dataTransfer.setData("text/plain", String(t))), e.style.opacity = "0.5";
    }), e.addEventListener("dragend", () => {
      e.style.opacity = "1";
    }), e.addEventListener("dragover", (r) => {
      r.preventDefault(), r.dataTransfer && (r.dataTransfer.dropEffect = "move");
    }), e.addEventListener("drop", (r) => {
      if (r.preventDefault(), r.dataTransfer && this.onDragDrop) {
        const s = parseInt(r.dataTransfer.getData("text/plain")), a = t;
        s !== a && this.onDragDrop(s, a);
      }
    });
  }
  /**
   * Update the dataset
   * @param data New dataset
   */
  setData(e) {
    this.data = e, this.scrollTop = 0, this.wrapper && (this.wrapper.scrollTop = 0), this.render();
  }
  /**
   * Update row height
   * @param height New row height in pixels
   */
  updateRowHeight(e) {
    this.rowHeight = e, this.render();
  }
  /**
   * Scroll to a specific row index
   * @param index Row index to scroll to
   */
  scrollToIndex(e) {
    if (!this.wrapper) return;
    const t = e * this.rowHeight;
    this.wrapper.scrollTop = t, this.scrollTop = t, this.render();
  }
  /**
   * Scroll to the top
   */
  scrollToTop() {
    this.scrollToIndex(0);
  }
  /**
   * Mount the virtual scroller to a parent element
   * @param parentElement Parent element to mount to
   */
  mount(e) {
    this.wrapper && (e.innerHTML = "", e.appendChild(this.wrapper), this.render());
  }
  /**
   * Refresh the current view
   */
  refresh() {
    this.render();
  }
  /**
   * Get the current scroll position
   * @returns Current scroll top position
   */
  getScrollTop() {
    return this.scrollTop;
  }
  /**
   * Get the total number of rows
   * @returns Total row count
   */
  getRowCount() {
    return this.data.length;
  }
  /**
   * Get the current visible start and end indices
   * @returns Object with visibleStart and visibleEnd
   */
  getVisibleIndices() {
    return {
      visibleStart: this.visibleStart,
      visibleEnd: this.visibleEnd
    };
  }
  /**
   * Clean up resources and remove event listeners
   */
  destroy() {
    this.scrollRAF !== null && (cancelAnimationFrame(this.scrollRAF), this.scrollRAF = null), this.resizeObserver && (this.resizeObserver.disconnect(), this.resizeObserver = null), this.wrapper && this.wrapper.parentElement && this.wrapper.parentElement.removeChild(this.wrapper), this.wrapper = null, this.spacer = null, this.content = null, this.data = [];
  }
}
export {
  le as ConnectService,
  tt as FieldService,
  z as PerformanceConfig,
  et as PivotEngine,
  rt as VirtualScrollManager,
  T as WasmLoader,
  He as getWasmLoader
};
