// @mindfiredigital/pivothead-llm/dist/index.js
// src/ModelManager.ts
var ModelManager = class {
  engine = null;
  ready = false;
  checkWebGPU() {
    if (typeof navigator !== "undefined" && "gpu" in navigator) {
      return { webgpu: true, message: "WebGPU is available" };
    }
    return {
      webgpu: false,
      message: "WebGPU is not supported in this browser"
    };
  }
  async load(modelId, onProgress) {
    if (!this.checkWebGPU().webgpu) {
      throw new Error(
        "WebGPU is not supported in this browser. Cannot load the model."
      );
    }
    const { CreateMLCEngine } = await import('@mlc-ai/web-llm');
    this.engine = await CreateMLCEngine(modelId, {
      initProgressCallback: (report) => {
        let stage = "downloading";
        if (report.progress >= 1) {
          stage = "ready";
        } else if (report.text?.toLowerCase().includes("init")) {
          stage = "initializing";
        }
        onProgress({ progress: report.progress, text: report.text, stage });
      }
    });
    this.ready = true;
    onProgress({ progress: 1, text: "Model ready", stage: "ready" });
  }
  async unload() {
    if (this.engine?.unload) {
      await this.engine.unload();
    }
    this.engine = null;
    this.ready = false;
  }
  isReady() {
    return this.ready;
  }
  getEngine() {
    return this.engine;
  }
};

// src/PromptBuilder.ts
var PromptBuilder = class {
  build(context) {
    const lines = [];
    lines.push(
      "You are an AI assistant for a pivot table application. Your role is to help users interact with and analyze their data through natural language commands."
    );
    lines.push("");
    lines.push("## Available Fields");
    for (const field of context.fields) {
      const valuesStr = field.values && field.values.length > 0 ? ` (values: ${field.values.join(", ")})` : "";
      lines.push(`- ${field.name} [${field.type}]${valuesStr}`);
    }
    lines.push("");
    lines.push("## Current Pivot State");
    const state = context.currentState;
    lines.push(`- Group By: ${state.groupBy ?? "none"}`);
    lines.push(`- Sort By: ${state.sortBy ?? "none"}`);
    const filters = state.filters && Object.keys(state.filters).length > 0 ? JSON.stringify(state.filters) : "none";
    lines.push(`- Filters: ${filters}`);
    lines.push("");
    lines.push("## Sample Data (up to 5 rows)");
    if (context.sampleRows.length > 0) {
      const headers = Object.keys(context.sampleRows[0]);
      lines.push(`| ${headers.join(" | ")} |`);
      lines.push(`| ${headers.map(() => "---").join(" | ")} |`);
      for (const row of context.sampleRows.slice(0, 5)) {
        lines.push(`| ${headers.map((h) => String(row[h] ?? "")).join(" | ")} |`);
      }
    } else {
      lines.push("No sample data available.");
    }
    lines.push("");
    lines.push("## Current Pivot Output (top 10 rows)");
    if (context.pivotOutput.length > 0) {
      const headers = Object.keys(context.pivotOutput[0]);
      lines.push(`| ${headers.join(" | ")} |`);
      lines.push(`| ${headers.map(() => "---").join(" | ")} |`);
      for (const row of context.pivotOutput.slice(0, 10)) {
        lines.push(`| ${headers.map((h) => String(row[h] ?? "")).join(" | ")} |`);
      }
    } else {
      lines.push("No pivot output available.");
    }
    lines.push("");
    lines.push("## Instructions");
    lines.push(
      "Respond ONLY with a single valid JSON object \u2014 no text, explanation, or markdown outside the JSON. Use EXACTLY the field names shown below. Do not rename fields."
    );
    lines.push("");
    lines.push(
      "Available action schemas (pick EXACTLY one \u2014 no other types allowed):"
    );
    lines.push(
      '{ "type": "sort", "field": "<fieldName>", "direction": "asc"|"desc" }'
    );
    lines.push(
      '{ "type": "filter", "field": "<fieldName>", "operator": "equals"|"contains"|"greaterThan"|"lessThan", "value": <value> }'
    );
    lines.push('{ "type": "removeFilter", "field": "<fieldName>" }');
    lines.push('{ "type": "groupBy", "field": "<fieldName>" }');
    lines.push(
      '{ "type": "topN", "n": <number>, "measure": "<fieldName>", "order": "asc"|"desc" }'
    );
    lines.push(
      '{ "type": "aggregate", "field": "<fieldName>", "func": "sum"|"avg"|"count"|"min"|"max" }'
    );
    lines.push('{ "type": "resetAll" }');
    lines.push('{ "type": "export", "format": "csv"|"json"|"pdf" }');
    lines.push('{ "type": "switchTab", "tab": "table"|"analytics" }');
    lines.push('{ "type": "chartType", "chartType": "<chartType>" }');
    lines.push('{ "type": "answer", "text": "<answer text>" }');
    lines.push('{ "type": "clarify", "question": "<clarifying question>" }');
    lines.push(
      '{ "type": "style", "target": "row"|"column", "value": "<row value or column name>", "property": "backgroundColor"|"color"|"fontWeight"|"fontStyle"|"fontSize", "style": "<css value>" }'
    );
    lines.push('{ "type": "resetStyle" }');
    lines.push("");
    lines.push(
      'STYLING RULES:\n- "make X red" or "highlight X" \u2192 target:"row", value:"X", property:"backgroundColor", style:"red"\n- "make X text red" or "X text color red" \u2192 target:"row", value:"X", property:"color", style:"red"\n- "make X column red" or "colour X column" \u2192 target:"column", value:"X", property:"backgroundColor", style:"red"\n- "make X bold" \u2192 property:"fontWeight", style:"bold"\n- "make X italic" \u2192 property:"fontStyle", style:"italic"\n- "remove styles" or "clear styles" \u2192 type:"resetStyle"\n- Color words: red, blue, green, yellow, orange, pink, purple, lightblue, lightgreen, lightyellow, #hexcode\n\nOTHER RULES:\n- If the request is about data (question, calculation, summary) \u2192 use "answer" with the computed result as text.\n- NEVER invent a new type. Only the types listed above are valid.\n- Do not add any text outside the JSON object.'
    );
    return lines.join("\n");
  }
};

// src/constants.ts
var DEFAULT_MODEL = "Llama-3.2-3B-Instruct-q4f16_1-MLC";
var DEFAULT_MAX_HISTORY = 10;
var DEFAULT_SYSTEM_PROMPT = "You are a pivot table AI assistant. Respond only with valid JSON action objects.";
var SWITCH_TAB_EVENT = "pivothead:switchTab";
var VALID_ACTION_TYPES = /* @__PURE__ */ new Set([
  "filter",
  "removeFilter",
  "sort",
  "groupBy",
  "topN",
  "aggregate",
  "resetAll",
  "export",
  "switchTab",
  "chartType",
  "answer",
  "clarify",
  "error",
  "style",
  "resetStyle"
]);

// src/ActionParser.ts
var ActionParser = class {
  /** Extracts the first balanced JSON object from text, ignoring anything after it. */
  extractFirstJson(text) {
    let depth = 0;
    let start = -1;
    for (let i = 0; i < text.length; i++) {
      if (text[i] === "{") {
        if (start === -1) start = i;
        depth++;
      } else if (text[i] === "}") {
        depth--;
        if (depth === 0 && start !== -1) {
          return text.slice(start, i + 1);
        }
      }
    }
    return null;
  }
  parse(rawText) {
    try {
      const jsonStr = this.extractFirstJson(rawText);
      if (!jsonStr) {
        return {
          type: "error",
          message: "No JSON object found in LLM response"
        };
      }
      const parsed = JSON.parse(jsonStr);
      if (typeof parsed.type !== "string" || !VALID_ACTION_TYPES.has(parsed.type)) {
        return {
          type: "answer",
          text: `I'm not able to "${String(parsed.type)}" \u2014 that's not something I can do on this table. I can help you with: filtering rows, sorting, computing aggregations (sum / avg / count / min / max), switching tabs, changing the chart type, or resetting the table. What would you like to do?`
        };
      }
      return parsed;
    } catch (err) {
      return {
        type: "error",
        message: `Failed to parse LLM response: ${err instanceof Error ? err.message : String(err)}`
      };
    }
  }
};

// src/LLMEngine.ts
var LLMEngine = class {
  modelManager;
  promptBuilder;
  actionParser;
  model;
  maxHistory;
  history = [];
  context = null;
  constructor(options = {}) {
    this.modelManager = new ModelManager();
    this.promptBuilder = new PromptBuilder();
    this.actionParser = new ActionParser();
    this.model = options.model ?? DEFAULT_MODEL;
    this.maxHistory = options.maxHistory ?? DEFAULT_MAX_HISTORY;
    if (options.onCapability) {
      const report = this.modelManager.checkWebGPU();
      options.onCapability(report);
    }
  }
  isReady() {
    return this.modelManager.isReady();
  }
  async load(onProgress) {
    await this.modelManager.load(this.model, onProgress ?? (() => {
    }));
  }
  async unload() {
    await this.modelManager.unload();
  }
  setContext(context) {
    this.context = context;
  }
  clearHistory() {
    this.history = [];
  }
  async query(userMessage) {
    if (!this.modelManager.isReady()) {
      throw new Error("Model is not ready. Call load() first.");
    }
    const messages = this.buildMessages(userMessage);
    const engine = this.modelManager.getEngine();
    const response = await engine.chat.completions.create({
      messages,
      stream: false
    });
    const assistantContent = response.choices[0]?.message?.content ?? "";
    this.appendHistory(userMessage, assistantContent);
    return this.actionParser.parse(assistantContent);
  }
  async *queryStream(userMessage, signal) {
    if (!this.modelManager.isReady()) {
      throw new Error("Model is not ready. Call load() first.");
    }
    const messages = this.buildMessages(userMessage);
    const engine = this.modelManager.getEngine();
    const stream = await engine.chat.completions.create({
      messages,
      stream: true
    });
    let fullContent = "";
    for await (const chunk of stream) {
      if (signal?.aborted) break;
      const delta = chunk.choices[0]?.delta?.content ?? "";
      if (delta) {
        fullContent += delta;
        yield delta;
      }
    }
    this.appendHistory(userMessage, fullContent);
  }
  buildMessages(userMessage) {
    const systemPrompt = this.context ? this.promptBuilder.build(this.context) : DEFAULT_SYSTEM_PROMPT;
    const recentHistory = this.history.slice(-(this.maxHistory * 2));
    return [
      { role: "system", content: systemPrompt },
      ...recentHistory,
      { role: "user", content: userMessage }
    ];
  }
  appendHistory(userMessage, assistantContent) {
    this.history.push({ role: "user", content: userMessage });
    this.history.push({ role: "assistant", content: assistantContent });
    if (this.history.length > this.maxHistory * 2) {
      this.history = this.history.slice(-(this.maxHistory * 2));
    }
  }
};

// src/ActionExecutor.ts
var ActionExecutor = class {
  pivotEngine;
  chartEngine;
  onActionApplied;
  onError;
  constructor(options) {
    this.pivotEngine = options.pivotEngine;
    this.chartEngine = options.chartEngine;
    this.onActionApplied = options.onActionApplied ?? (() => {
    });
    this.onError = options.onError ?? (() => {
    });
  }
  async execute(action) {
    try {
      switch (action.type) {
        case "filter":
          this.pivotEngine.applyFilter?.({
            field: action.field,
            operator: action.operator,
            value: action.value
          });
          this.onActionApplied(action, {
            success: true,
            description: `Filtered ${action.field} ${action.operator} ${String(action.value)}`
          });
          break;
        case "removeFilter":
          this.pivotEngine.removeFilter?.(action.field);
          this.onActionApplied(action, {
            success: true,
            description: `Removed filter on ${action.field}`
          });
          break;
        case "sort":
          this.pivotEngine.sortData?.(action.field, action.direction);
          this.onActionApplied(action, {
            success: true,
            description: `Sorted by ${action.field} (${action.direction})`
          });
          break;
        case "groupBy":
          this.pivotEngine.groupData?.(action.field);
          this.onActionApplied(action, {
            success: true,
            description: `Grouped by ${action.field}`
          });
          break;
        case "topN":
          this.pivotEngine.applyTopN?.(action.n, action.measure, action.order);
          this.onActionApplied(action, {
            success: true,
            description: `Showing top ${action.n} by ${action.measure}`
          });
          break;
        case "aggregate":
          this.pivotEngine.setAggregation?.(action.field, action.func);
          this.onActionApplied(action, {
            success: true,
            description: `Aggregation for ${action.field} set to ${action.func}`
          });
          break;
        case "resetAll":
          this.pivotEngine.reset?.();
          this.onActionApplied(action, {
            success: true,
            description: "Table reset to default state"
          });
          break;
        case "export":
          this.pivotEngine.export?.(action.format);
          this.onActionApplied(action, {
            success: true,
            description: `Exporting as ${action.format}`
          });
          break;
        case "switchTab":
          window.dispatchEvent(
            new CustomEvent(SWITCH_TAB_EVENT, {
              detail: { tab: action.tab }
            })
          );
          this.onActionApplied(action, {
            success: true,
            description: `Switched to ${action.tab} tab`
          });
          break;
        case "chartType":
          this.chartEngine?.updateChartType?.(action.chartType);
          this.onActionApplied(action, {
            success: true,
            description: `Chart type changed to ${action.chartType}`
          });
          break;
        case "answer":
          this.onActionApplied(action, {
            success: true,
            description: action.text
          });
          break;
        case "clarify":
          this.onActionApplied(action, {
            success: true,
            description: action.question
          });
          break;
        case "style":
          this.pivotEngine.applyStyle?.(
            action.target,
            action.value,
            action.property,
            action.style
          );
          this.onActionApplied(action, {
            success: true,
            description: `Style applied: ${action.target} "${action.value}" \u2192 ${action.property}: ${action.style}`
          });
          break;
        case "resetStyle":
          this.pivotEngine.resetStyle?.();
          this.onActionApplied(action, {
            success: true,
            description: "All styles cleared"
          });
          break;
        case "error":
          this.onError(action, new Error(action.message));
          break;
      }
    } catch (err) {
      this.onError(action, err instanceof Error ? err : new Error(String(err)));
    }
  }
};

// src/ContextBuilder.ts
var ContextBuilder = class {
  static fromRawData(rawData, options = {}) {
    if (!rawData.length) {
      return { fields: [], sampleRows: [], pivotOutput: [], currentState: {} };
    }
    const firstRow = rawData[0];
    const allKeys = Object.keys(firstRow);
    const fields = allKeys.map((name) => {
      const val = firstRow[name];
      const type = typeof val === "number" ? "number" : "string";
      if (type === "string") {
        const unique = [
          ...new Set(rawData.map((r) => String(r[name])).filter(Boolean))
        ];
        if (unique.length <= 20) return { name, type, values: unique };
      }
      return { name, type };
    });
    const rowField = options.groupConfig?.rowFields?.[0] ?? null;
    const colField = options.groupConfig?.columnFields?.[0] ?? null;
    const numericFields = fields.filter((f) => f.type === "number").map((f) => f.name);
    let pivotOutput = [];
    if (rowField) {
      const groupMap = /* @__PURE__ */ new Map();
      rawData.forEach((row) => {
        const key = colField ? `${String(row[rowField])}||${String(row[colField])}` : String(row[rowField]);
        if (!groupMap.has(key)) {
          const entry2 = { [rowField]: row[rowField] };
          if (colField) entry2[colField] = row[colField];
          numericFields.forEach((f) => entry2[`${f}_sum`] = 0);
          entry2["_count"] = 0;
          groupMap.set(key, entry2);
        }
        const entry = groupMap.get(key);
        numericFields.forEach(
          (f) => entry[`${f}_sum`] = (Number(entry[`${f}_sum`]) || 0) + (Number(row[f]) || 0)
        );
        entry["_count"] = (Number(entry["_count"]) || 0) + 1;
      });
      pivotOutput = [...groupMap.values()];
      pivotOutput.forEach((row) => {
        numericFields.forEach((f) => {
          const key = `${f}_sum`;
          if (typeof row[key] === "number") {
            row[key] = Math.round(row[key] * 100) / 100;
          }
        });
      });
    }
    const sortConfig = Array.isArray(options.sortConfig) ? options.sortConfig[0] : options.sortConfig;
    return {
      fields,
      sampleRows: rawData.slice(0, 5),
      pivotOutput: pivotOutput.slice(0, 10),
      currentState: {
        groupBy: rowField ?? void 0,
        sortBy: sortConfig?.["field"],
        filters: options.filters ?? {}
      }
    };
  }
};

// src/LLMAssistant.ts
var LLMAssistant = class {
  llmEngine;
  pivotEngine;
  chartEngine;
  onMessage;
  constructor(options) {
    this.pivotEngine = options.engine;
    this.chartEngine = options.chartEngine;
    this.onMessage = options.onMessage;
    this.llmEngine = new LLMEngine({
      model: options.model,
      maxHistory: options.maxHistory,
      onCapability: options.onCapability
    });
  }
  /** Whether the model is loaded and ready to accept queries */
  isReady() {
    return this.llmEngine.isReady();
  }
  /**
   * Download and initialise the LLM model (~1.5 GB, cached by the browser after first load).
   * @param onProgress Optional callback for download/init progress (0–1).
   */
  async load(onProgress) {
    await this.llmEngine.load(onProgress);
  }
  /** Unload the model and free GPU memory */
  async unload() {
    await this.llmEngine.unload();
  }
  /**
   * Send a natural-language query. The assistant will:
   * 1. Auto-build context from the current engine state
   * 2. Query the LLM
   * 3. Execute the resulting action on the engine
   * 4. Call onMessage with the result
   */
  async query(text) {
    const state = this.pivotEngine.getState();
    const rawData = state.rawData ?? [];
    const context = ContextBuilder.fromRawData(rawData, {
      groupConfig: state.groupConfig,
      sortConfig: state.sortConfig,
      filters: state.filters
    });
    this.llmEngine.setContext(context);
    const action = await this.llmEngine.query(text);
    await this.buildExecutor(rawData, context, text.toLowerCase()).execute(
      action
    );
  }
  /** Clear LLM conversation history */
  clearHistory() {
    this.llmEngine.clearHistory();
  }
  buildExecutor(rawData, context, queryText) {
    const engine = this.pivotEngine;
    const chartEngine = this.chartEngine;
    const onMessage = this.onMessage;
    function computeAgg(values, fn) {
      if (!values.length) return 0;
      switch (fn.toLowerCase()) {
        case "sum":
          return values.reduce((a, b) => a + b, 0);
        case "avg":
        case "average":
          return values.reduce((a, b) => a + b, 0) / values.length;
        case "count":
          return values.length;
        case "min":
          return Math.min(...values);
        case "max":
          return Math.max(...values);
        default:
          return values.reduce((a, b) => a + b, 0);
      }
    }
    const fmt = (n) => Math.round(n * 100) / 100;
    return new ActionExecutor({
      pivotEngine: {
        applyFilter: ({ field, operator, value }) => {
          const current = engine.getFilterState?.() ?? [];
          const without = current.filter((f) => f.field !== field);
          engine.applyFilters([...without, { field, operator, value }]);
        },
        removeFilter: (field) => {
          const current = engine.getFilterState?.() ?? [];
          engine.applyFilters(current.filter((f) => f.field !== field));
        },
        sortData: (field, direction) => {
          engine.sort(field, direction);
        },
        reset: () => engine.reset(),
        /**
         * Computes the aggregation from raw data and delivers the result via onMessage.
         * Also narrows the breakdown to groups mentioned in the query.
         */
        setAggregation: (field, func) => {
          if (!rawData.length || !field) {
            onMessage(
              "error",
              `Cannot compute ${func}(${field}): no data available.`
            );
            return;
          }
          const allValues = rawData.map((r) => Number(r[field])).filter((v) => !isNaN(v));
          const grandTotal = fmt(computeAgg(allValues, func));
          const rowField = context.currentState.groupBy;
          const colField = engine.getState().groupConfig?.columnFields?.[0];
          if (!rowField) {
            onMessage(
              "assistant",
              `${func.toUpperCase()}(${field}) = ${grandTotal.toLocaleString()}`
            );
            return;
          }
          const groupMap = /* @__PURE__ */ new Map();
          rawData.forEach((row) => {
            const rv = String(row[rowField]);
            const cv = colField ? String(row[colField]) : null;
            const key = cv ? `${rv} / ${cv}` : rv;
            const val = Number(row[field]);
            if (!isNaN(val)) {
              if (!groupMap.has(key)) groupMap.set(key, []);
              groupMap.get(key).push(val);
            }
          });
          const stopWords = /* @__PURE__ */ new Set([
            "what",
            "is",
            "the",
            "of",
            "for",
            "and",
            "in",
            "a",
            "an",
            "sum",
            "avg",
            "count",
            "min",
            "max",
            "total",
            "average",
            "show",
            "me",
            "give",
            "tell",
            "how",
            "much",
            "many",
            "are",
            "by",
            "to",
            "or"
          ]);
          const queryWords = queryText.replace(/[^a-z0-9\s]/g, " ").split(/\s+/).filter(
            (w) => w.length > 1 && !stopWords.has(w) && w !== field.toLowerCase()
          );
          const rowValues = [...new Set(rawData.map((r) => String(r[rowField])))];
          const colValues = colField ? [...new Set(rawData.map((r) => String(r[colField])))] : [];
          const matchedRowVals = rowValues.filter(
            (v) => queryWords.some((w) => v.toLowerCase().includes(w))
          );
          const matchedColVals = colValues.filter(
            (v) => queryWords.some((w) => v.toLowerCase().includes(w))
          );
          const relevantEntries = matchedRowVals.length > 0 || matchedColVals.length > 0 ? [...groupMap.entries()].filter(([k]) => {
            const parts = k.split(" / ");
            const rv = parts[0];
            const cv = parts[1] ?? null;
            const rowOk = matchedRowVals.length === 0 || matchedRowVals.some(
              (v) => v.toLowerCase() === rv.toLowerCase()
            );
            const colOk = matchedColVals.length === 0 || cv !== null && matchedColVals.some(
              (v) => v.toLowerCase() === cv.toLowerCase()
            );
            return rowOk && colOk;
          }) : [];
          if (relevantEntries.length > 0) {
            const filteredValues = relevantEntries.flatMap(([, vals]) => vals);
            const filteredTotal = fmt(computeAgg(filteredValues, func));
            const lines = relevantEntries.map(
              ([k, vals]) => `  ${k}: ${fmt(computeAgg(vals, func)).toLocaleString()}`
            ).join("\n");
            const groupLabel = colField ? `${rowField} \xD7 ${colField}` : rowField;
            onMessage(
              "assistant",
              `${func.toUpperCase()}(${field}) = ${filteredTotal.toLocaleString()}

Breakdown by ${groupLabel}:
${lines}`
            );
          } else {
            const lines = [...groupMap.entries()].map(
              ([k, vals]) => `  ${k}: ${fmt(computeAgg(vals, func)).toLocaleString()}`
            ).join("\n");
            const groupLabel = colField ? `${rowField} \xD7 ${colField}` : rowField;
            onMessage(
              "assistant",
              `${func.toUpperCase()}(${field}) = ${grandTotal.toLocaleString()}

Breakdown by ${groupLabel}:
${lines}`
            );
          }
        },
        applyStyle: (_target, _value, _property, _style) => {
        },
        resetStyle: () => {
        },
        groupData: (_field) => {
          onMessage(
            "assistant",
            "groupBy requires rebuilding the pivot config \u2014 not supported via LLM in this version."
          );
        },
        applyTopN: (_n, _measure, _order) => {
          onMessage("assistant", "topN is not yet supported.");
        },
        export: (format) => {
          onMessage(
            "assistant",
            `Export as ${format}: handle this in your app via the onMessage callback.`
          );
        }
      },
      chartEngine: chartEngine ? { updateChartType: (t) => chartEngine.updateChartType?.(t) } : void 0,
      onActionApplied: (_action, result) => {
        onMessage("assistant", `\u2713 ${result.description}`);
      },
      onError: (_action, error) => {
        onMessage("assistant", error.message);
      }
    });
  }
};

export { ActionExecutor, ContextBuilder, LLMAssistant, LLMEngine };
